<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">

<head>

<meta charset="utf-8" />
<meta name="generator" content="quarto-1.3.433" />

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />

<meta name="author" content="J. Serizay, O. Ashenberg &amp; F. Almeida-Silva" />

<title>Single-cell RNAseq analysis with R/Bioconductor – 9  Lab 5: Dimension reduction, clustering and annotation</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<!-- htmldependencies:E3FAD763 -->
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body>

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn"
      data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" 
      aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation"
      onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <h1 class="quarto-secondary-nav-title"></h1>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" 
      aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation"
      onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="/">
      Single-cell RNAseq analysis with R/Bioconductor
      </a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/js2264/scRNAseq_Physalia_2025/" rel="" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/program.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Program</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/Rstudio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">RStudio</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/prerequisites.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prerequisites</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Day 1</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/day1/Lecture1_introduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>1</span>  <span class='chapter-title'>Lecture 1 - Introduction to scRNAseq analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/day1/Lecture2_Bcl2matrix.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>2</span>  <span class='chapter-title'>Lecture 2 - From sequencing reads to expression matrices</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/day1/Lab1_Intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>3</span>  <span class='chapter-title'>Lab 1: Familiarizing yourself with the course AWS instance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/day1/Lab2_processingreads.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>4</span>  <span class='chapter-title'>Lab 2: From .bcl to count matrix</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Day 2</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/day2/Lecture3_qualitycontrol.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>5</span>  <span class='chapter-title'>Lecture 3 - Quality control for scRNA-Seq data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/day2/Lab3_Rbioc.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>6</span>  <span class='chapter-title'>Lab 3: Introduction to R/Bioconductor</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/day2/Lab4_data_wrangling_scRNAseq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>7</span>  <span class='chapter-title'>Lab 4 - Single-cell RNA-seq data wrangling</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Day 3</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/day3/Lecture4_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>8</span>  <span class='chapter-title'>Lecture 4 - Identifying cell populations</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/day3/Lab5_clustering.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class='chapter-number'>9</span>  <span class='chapter-title'>Lab 5: Dimension reduction, clustering and annotation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/day3/Lecture5_batchcorrection.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>10</span>  <span class='chapter-title'>Lecture 5 - Data integration and batch effect correction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/day3/Lab6_batch_correction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>11</span>  <span class='chapter-title'>Lab 6: Batch correction</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">Day 4</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/day4/Lecture6_ATAC.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>12</span>  <span class='chapter-title'>Lecture 6 - Advances in single-cell genomics: the epigenome</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/day4/Lab7_atac-seq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>13</span>  <span class='chapter-title'>Lab 7: Single-cell ATAC-seq analysis workflow</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/day4/Lecture7_pseudotime.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>14</span>  <span class='chapter-title'>Lecture 7 - Trajectories and pseudotimes</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/day4/Lab8_pseudotime.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>15</span>  <span class='chapter-title'>Lab 8: Pseudotime analyses</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Day 5</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/day5/Lecture8_spatial-transcriptomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class='chapter-number'>16</span>  <span class='chapter-title'>Lecture 8 - Advances in single-cell genomics: spatial transcriptomics</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="/content/extra.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Extra resources</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" ></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <div id="quarto-toc-target"></div>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">9</span>  <span class="chapter-title">Lab 5: Dimension reduction, clustering and annotation</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>
<nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#dimensional-reduction-for-clustering" id="toc-dimensional-reduction-for-clustering"><span class="header-section-number">9.1</span> Dimensional reduction for clustering</a>
  <ul>
  <li><a href="#preparing-dataset" id="toc-preparing-dataset"><span class="header-section-number">9.1.1</span> Preparing dataset</a></li>
  <li><a href="#normalize-counts-using-scran" id="toc-normalize-counts-using-scran"><span class="header-section-number">9.1.2</span> Normalize counts using <code>scran</code></a></li>
  <li><a href="#feature-selection" id="toc-feature-selection"><span class="header-section-number">9.1.3</span> Feature selection</a></li>
  <li><a href="#pca-on-filtered-dataset" id="toc-pca-on-filtered-dataset"><span class="header-section-number">9.1.4</span> PCA on filtered dataset</a></li>
  </ul></li>
  <li><a href="#clustering" id="toc-clustering"><span class="header-section-number">9.2</span> Clustering</a>
  <ul>
  <li><a href="#clustering-algorithms" id="toc-clustering-algorithms"><span class="header-section-number">9.2.1</span> Clustering algorithms</a></li>
  <li><a href="#dimensional-reduction-for-clustering-visualization" id="toc-dimensional-reduction-for-clustering-visualization"><span class="header-section-number">9.2.2</span> Dimensional reduction for clustering visualization</a></li>
  <li><a href="#bonus-for-the-pros-of-clustering-compare-different-clustering-approaches" id="toc-bonus-for-the-pros-of-clustering-compare-different-clustering-approaches"><span class="header-section-number">9.2.3</span> BONUS For the pros of clustering… Compare different clustering approaches</a></li>
  </ul></li>
  <li><a href="#cell-annotation" id="toc-cell-annotation"><span class="header-section-number">9.3</span> Cell annotation</a>
  <ul>
  <li><a href="#find-marker-genes" id="toc-find-marker-genes"><span class="header-section-number">9.3.1</span> Find marker genes</a></li>
  <li><a href="#automated-cell-annotation" id="toc-automated-cell-annotation"><span class="header-section-number">9.3.2</span> Automated cell annotation</a></li>
  </ul></li>
  <li><a href="#bonus" id="toc-bonus"><span class="header-section-number">9.4</span> Bonus</a></li>
  <li><a href="#acknowledgements" id="toc-acknowledgements"><span class="header-section-number">9.5</span> Acknowledgements</a></li>
  <li><a href="#session-info" id="toc-session-info"><span class="header-section-number">9.6</span> Session info</a></li>
  </ul>
</nav>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Notes
</div>
</div>
<div class="callout-body-container callout-body">
<p>The estimated time for this lab is around 1h20.</p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Aims
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Learn how to perform dimension reduction on scRNAseq data</li>
<li>Understand the importance of feature selection</li>
<li>Learn how to perform clustering on scRNAseq data</li>
<li>Learn how to (automatically) annotate cells based on clustering results</li>
</ul>
</div>
</div>
<section id="dimensional-reduction-for-clustering" class="level2" data-number="9.1">
<h2 data-number="9.1"><span class="header-section-number">9.1</span> Dimensional reduction for clustering</h2>
<section id="preparing-dataset" class="level3" data-number="9.1.1">
<h3 data-number="9.1.1"><span class="header-section-number">9.1.1</span> Preparing dataset</h3>
<p>We will prepare scRNAseq data from a PBMC run, provided by 10X and hosted by <code>Bioconductor</code> as a package.</p>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Which package from <code>Bioconductor</code> gives streamlined access to PBMC scRNAseq dataset from 10X Genomics?</p>
<p>What does the object contain (type of data, number of cells, batches, organism, …)? Can you get the same data from somewhere else?</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb1"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div class="sourceCode" id="cb3"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SingleCellExperiment)</span></code></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: SummarizedExperiment
Loading required package: MatrixGenerics
Loading required package: matrixStats

Attaching package: &#39;matrixStats&#39;

The following object is masked from &#39;package:dplyr&#39;:

    count


Attaching package: &#39;MatrixGenerics&#39;

The following objects are masked from &#39;package:matrixStats&#39;:

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse, colCounts, colCummaxs, colCummins, colCumprods, colCumsums, colDiffs, colIQRDiffs, colIQRs, colLogSumExps,
    colMadDiffs, colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats, colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds, colSums2, colTabulates,
    colVarDiffs, colVars, colWeightedMads, colWeightedMeans, colWeightedMedians, colWeightedSds, colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet, rowCollapse,
    rowCounts, rowCummaxs, rowCummins, rowCumprods, rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps, rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks, rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars, rowWeightedMads, rowWeightedMeans,
    rowWeightedMedians, rowWeightedSds, rowWeightedVars

Loading required package: GenomicRanges
Loading required package: stats4
Loading required package: BiocGenerics

Attaching package: &#39;BiocGenerics&#39;

The following objects are masked from &#39;package:lubridate&#39;:

    intersect, setdiff, union

The following objects are masked from &#39;package:dplyr&#39;:

    combine, intersect, setdiff, union

The following objects are masked from &#39;package:stats&#39;:

    IQR, mad, sd, var, xtabs

The following objects are masked from &#39;package:base&#39;:

    anyDuplicated, aperm, append, as.data.frame, basename, cbind, colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep, grepl, intersect, is.unsorted,
    lapply, Map, mapply, match, mget, order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind, Reduce, rownames, sapply, setdiff, table, tapply, union, unique,
    unsplit, which.max, which.min

Loading required package: S4Vectors

Attaching package: &#39;S4Vectors&#39;

The following objects are masked from &#39;package:lubridate&#39;:

    second, second&lt;-

The following objects are masked from &#39;package:dplyr&#39;:

    first, rename

The following object is masked from &#39;package:tidyr&#39;:

    expand

The following object is masked from &#39;package:utils&#39;:

    findMatches

The following objects are masked from &#39;package:base&#39;:

    expand.grid, I, unname

Loading required package: IRanges

Attaching package: &#39;IRanges&#39;

The following object is masked from &#39;package:lubridate&#39;:

    %within%

The following objects are masked from &#39;package:dplyr&#39;:

    collapse, desc, slice

The following object is masked from &#39;package:purrr&#39;:

    reduce

Loading required package: GenomeInfoDb
Loading required package: Biobase
Welcome to Bioconductor

    Vignettes contain introductory material; view with &#39;browseVignettes()&#39;. To cite Bioconductor, see &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.


Attaching package: &#39;Biobase&#39;

The following object is masked from &#39;package:MatrixGenerics&#39;:

    rowMedians

The following objects are masked from &#39;package:matrixStats&#39;:

    anyMissing, rowMedians</code></pre>
</div>
<div class="sourceCode" id="cb5"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> TENxPBMCData<span class="sc">::</span><span class="fu">TENxPBMCData</span>(<span class="st">&#39;pbmc4k&#39;</span>)</span></code></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>see ?TENxPBMCData and browseVignettes(&#39;TENxPBMCData&#39;) for documentation
loading from cache</code></pre>
</div>
<div class="sourceCode" id="cb7"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(sce) <span class="ot">&lt;-</span> scuttle<span class="sc">::</span><span class="fu">uniquifyFeatureNames</span>(<span class="fu">rowData</span>(sce)<span class="sc">$</span>ENSEMBL_ID, <span class="fu">rowData</span>(sce)<span class="sc">$</span>Symbol_TENx)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>sce</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SingleCellExperiment 
dim: 33694 4340 
metadata(0):
assays(1): counts
rownames(33694): RP11-34P13.3 FAM138A ... AC213203.1 FAM231B
rowData names(3): ENSEMBL_ID Symbol_TENx Symbol
colnames: NULL
colData names(11): Sample Barcode ... Individual Date_published
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
<div class="sourceCode" id="cb9"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 33694 rows and 3 columns
                  ENSEMBL_ID  Symbol_TENx       Symbol
                 &lt;character&gt;  &lt;character&gt;  &lt;character&gt;
RP11-34P13.3 ENSG00000243485 RP11-34P13.3           NA
FAM138A      ENSG00000237613      FAM138A      FAM138A
OR4F5        ENSG00000186092        OR4F5        OR4F5
RP11-34P13.7 ENSG00000238009 RP11-34P13.7 LOC100996442
RP11-34P13.8 ENSG00000239945 RP11-34P13.8           NA
...                      ...          ...          ...
AC233755.2   ENSG00000277856   AC233755.2           NA
AC233755.1   ENSG00000275063   AC233755.1 LOC102723407
AC240274.1   ENSG00000271254   AC240274.1 LOC102724250
AC213203.1   ENSG00000277475   AC213203.1           NA
FAM231B      ENSG00000268674      FAM231B      FAM231C</code></pre>
</div>
<div class="sourceCode" id="cb11"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(sce)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 4340 rows and 11 columns
          Sample            Barcode         Sequence   Library Cell_ranger_version Tissue_status Barcode_type   Chemistry Sequence_platform    Individual Date_published
     &lt;character&gt;        &lt;character&gt;      &lt;character&gt; &lt;integer&gt;         &lt;character&gt;   &lt;character&gt;  &lt;character&gt; &lt;character&gt;       &lt;character&gt;   &lt;character&gt;    &lt;character&gt;
1         pbmc4k AAACCTGAGAAGGCCT-1 AAACCTGAGAAGGCCT         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08
2         pbmc4k AAACCTGAGACAGACC-1 AAACCTGAGACAGACC         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08
3         pbmc4k AAACCTGAGATAGTCA-1 AAACCTGAGATAGTCA         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08
4         pbmc4k AAACCTGAGCGCCTCA-1 AAACCTGAGCGCCTCA         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08
5         pbmc4k AAACCTGAGGCATGGT-1 AAACCTGAGGCATGGT         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08
...          ...                ...              ...       ...                 ...           ...          ...         ...               ...           ...            ...
4336      pbmc4k TTTGGTTTCGCTAGCG-1 TTTGGTTTCGCTAGCG         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08
4337      pbmc4k TTTGTCACACTTAACG-1 TTTGTCACACTTAACG         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08
4338      pbmc4k TTTGTCACAGGTCCAC-1 TTTGTCACAGGTCCAC         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08
4339      pbmc4k TTTGTCAGTTAAGACA-1 TTTGTCAGTTAAGACA         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08
4340      pbmc4k TTTGTCATCCCAAGAT-1 TTTGTCATCCCAAGAT         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08</code></pre>
</div>
<div class="sourceCode" id="cb13"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(sce<span class="sc">$</span>Library)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   1 
4340 </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="normalize-counts-using-scran" class="level3" data-number="9.1.2">
<h3 data-number="9.1.2"><span class="header-section-number">9.1.2</span> Normalize counts using <code>scran</code></h3>
<p>Just like in bulk high-throughput sequencing experiments, scRNAseq counts have to be normalized to the sequencing depth for each cell. We can define the library size as the total sum of counts across all genes for each cell, the expected value of which is assumed to scale with any cell-specific biases. However, this relies on the assumption that within the entire dataset, most genes are non-differentially expressed and expressed roughly within the same range. Depending on the set up of the scRNAseq experiment, this can be entirely false. To avoid relying on this hypothesis, we can (1) quickly pre-cluster cells, then (2) normalize cells using their library size factor separately in each cluster, then (3) rescaling size factors so that they are comparable across clusters.</p>
<p>All of this can be done very simply using the combo <code>quickCluster() + computeSumFactors() + logNormCounts()</code> from <code>scran/scuttle</code> packages.</p>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>What is the purpose of the <code>quickCluster()</code> function? What does it return?</p>
<p>What is the purpose of the <code>computeSumFactors()</code> function? What does it return?</p>
<p>What is the purpose of the <code>logNormCounts()</code> function? What does it return?</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb15"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>clusters <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">quickCluster</span>(sce)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(clusters)</span></code></pre></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>clusters
  1   2   3   4   5   6   7   8   9  10  11 
588 361 264 161 866 365 531 536 360 207 101 </code></pre>
</div>
<div class="sourceCode" id="cb17"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">computeSumFactors</span>(sce, <span class="at">cluster =</span> clusters)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(sce)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 4340 rows and 12 columns
          Sample            Barcode         Sequence   Library Cell_ranger_version Tissue_status Barcode_type   Chemistry Sequence_platform    Individual Date_published sizeFactor
     &lt;character&gt;        &lt;character&gt;      &lt;character&gt; &lt;integer&gt;         &lt;character&gt;   &lt;character&gt;  &lt;character&gt; &lt;character&gt;       &lt;character&gt;   &lt;character&gt;    &lt;character&gt;  &lt;numeric&gt;
1         pbmc4k AAACCTGAGAAGGCCT-1 AAACCTGAGAAGGCCT         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08   0.403246
2         pbmc4k AAACCTGAGACAGACC-1 AAACCTGAGACAGACC         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08   0.739777
3         pbmc4k AAACCTGAGATAGTCA-1 AAACCTGAGATAGTCA         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08   0.429493
4         pbmc4k AAACCTGAGCGCCTCA-1 AAACCTGAGCGCCTCA         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08   0.527001
5         pbmc4k AAACCTGAGGCATGGT-1 AAACCTGAGGCATGGT         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08   0.600344
...          ...                ...              ...       ...                 ...           ...          ...         ...               ...           ...            ...        ...
4336      pbmc4k TTTGGTTTCGCTAGCG-1 TTTGGTTTCGCTAGCG         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08   1.422566
4337      pbmc4k TTTGTCACACTTAACG-1 TTTGTCACACTTAACG         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08   0.947146
4338      pbmc4k TTTGTCACAGGTCCAC-1 TTTGTCACAGGTCCAC         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08   1.962193
4339      pbmc4k TTTGTCAGTTAAGACA-1 TTTGTCAGTTAAGACA         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08   0.615199
4340      pbmc4k TTTGTCATCCCAAGAT-1 TTTGTCATCCCAAGAT         1                v2.1            NA     Chromium Chromium_v2         Hiseq4000 HealthyDonor1     2017-11-08   0.771550</code></pre>
</div>
<div class="sourceCode" id="cb19"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> scuttle<span class="sc">::</span><span class="fu">logNormCounts</span>(sce)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">assays</span>(sce)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of length 2
names(2): counts logcounts</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="feature-selection" class="level3" data-number="9.1.3">
<h3 data-number="9.1.3"><span class="header-section-number">9.1.3</span> Feature selection</h3>
<p>We often use scRNAseq data in exploratory analyses to characterize heterogeneity across cells. Procedures like clustering and dimensionality reduction compare cells based on their gene expression profiles. The choice of genes to include in this comparison may have a major impact on the performance of downstream methods. Ideally, one wants to only select genes that contain useful information about the biology of the system while removing genes that contain random noise. This aims to preserve interesting biological structure without the variance that obscures that structure.</p>
<p>The simplest approach to feature selection is to compute the variance of the log-normalized expression values, to select the most variable genes. Modelling of the mean-variance relationship can be achieved by the <code>modelGeneVar()</code> function from the <code>scran</code> package.</p>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Apply the <code>modelGeneVar()</code> function to the normalized counts. What does the returned object contain?</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb21"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>sce_filtered_variance <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">modelGeneVar</span>(sce)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>sce_filtered_variance</span></code></pre></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 33694 rows and 6 columns
                    mean       total        tech          bio   p.value       FDR
               &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;    &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
RP11-34P13.3 0.000000000 0.000000000 0.000000000  0.00000e+00       NaN       NaN
FAM138A      0.000000000 0.000000000 0.000000000  0.00000e+00       NaN       NaN
OR4F5        0.000000000 0.000000000 0.000000000  0.00000e+00       NaN       NaN
RP11-34P13.7 0.001973297 0.002003373 0.002046038 -4.26643e-05  0.549850  0.744658
RP11-34P13.8 0.000491705 0.000531807 0.000509831  2.19764e-05  0.397824  0.744658
...                  ...         ...         ...          ...       ...       ...
AC233755.2    0.00000000   0.0000000   0.0000000   0.00000000       NaN       NaN
AC233755.1    0.00000000   0.0000000   0.0000000   0.00000000       NaN       NaN
AC240274.1    0.00961767   0.0112683   0.0099722   0.00129611  0.217433  0.744658
AC213203.1    0.00000000   0.0000000   0.0000000   0.00000000       NaN       NaN
FAM231B       0.00000000   0.0000000   0.0000000   0.00000000       NaN       NaN</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Identify the top 10% most variable genes from this object.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-8-contents" aria-controls="callout-8" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-8" class="callout-8-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb23"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>HVGs <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">getTopHVGs</span>(sce_filtered_variance, <span class="at">prop =</span> <span class="fl">0.1</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>isHVG <span class="ot">&lt;-</span> <span class="fu">rownames</span>(sce) <span class="sc">%in%</span> HVGs</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rowData</span>(sce))</span></code></pre></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 4 columns
                   ENSEMBL_ID   Symbol_TENx       Symbol     isHVG
                  &lt;character&gt;   &lt;character&gt;  &lt;character&gt; &lt;logical&gt;
RP11-34P13.3  ENSG00000243485  RP11-34P13.3           NA     FALSE
FAM138A       ENSG00000237613       FAM138A      FAM138A     FALSE
OR4F5         ENSG00000186092         OR4F5        OR4F5     FALSE
RP11-34P13.7  ENSG00000238009  RP11-34P13.7 LOC100996442     FALSE
RP11-34P13.8  ENSG00000239945  RP11-34P13.8           NA     FALSE
RP11-34P13.14 ENSG00000239906 RP11-34P13.14           NA     FALSE</code></pre>
</div>
<div class="sourceCode" id="cb25"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">rowData</span>(sce)<span class="sc">$</span>isHVG)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
FALSE  TRUE 
32424  1270 </code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Visualize the mean-variance fit of the data, coloring points by whether they are in the top 10% most variable genes.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb27"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean =</span> <span class="fu">metadata</span>(sce_filtered_variance)<span class="sc">$</span>mean, </span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">var =</span> <span class="fu">metadata</span>(sce_filtered_variance)<span class="sc">$</span>var, </span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">trend =</span> <span class="fu">metadata</span>(sce_filtered_variance)<span class="sc">$</span><span class="fu">trend</span>(mean), </span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">HVG =</span> <span class="fu">rowData</span>(sce)<span class="sc">$</span>isHVG</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df) <span class="sc">+</span> </span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> mean, <span class="at">y =</span> var, <span class="at">col =</span> HVG), <span class="at">alpha =</span> <span class="fl">0.4</span>) <span class="sc">+</span> </span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>(<span class="fu">aes</span>(<span class="at">x =</span> mean, <span class="at">y =</span> trend), <span class="at">col =</span> <span class="st">&#39;darkred&#39;</span>) <span class="sc">+</span></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>() <span class="sc">+</span> </span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&#39;Gene mean exp. (norm.)&#39;</span>, <span class="at">y =</span> <span class="st">&#39;Gene exp. variance&#39;</span>)</span></code></pre></div>
</div>
<div class="cell-output-display">
<p><img src="Lab5_clustering_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid" width="672" /></p>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="pca-on-filtered-dataset" class="level3" data-number="9.1.4">
<h3 data-number="9.1.4"><span class="header-section-number">9.1.4</span> PCA on filtered dataset</h3>
<p>We now have normalized counts filtered for the top 500 genes varying with the greatest biological significance.<br />
Still, that represents a 500 x nCells (~8,000) dataset (each row being a feature). This is still too big to reliably use in standard clustering approaches. We can further compress the dataset. The most widely used approach is <code>PCA</code>: it computes a small number of “components” (typically 5-50) optimally summarizing the variability of the whole dataset, while retaining linearity of the underlying numerical data and being computationallt quite efficient.</p>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Leverage the <code>scater</code> package to compute a <code>PCA</code> embedding of the filtered data by taking into account the technical variability.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-11-contents" aria-controls="callout-11" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-11" class="callout-11-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb28"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">denoisePCA</span>(</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>    sce, </span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">technical =</span> sce_filtered_variance, </span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">subset.row =</span> HVGs, </span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">min.rank =</span> <span class="dv">15</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(sce, <span class="st">&#39;PCA&#39;</span>, <span class="at">colour_by =</span> <span class="st">&#39;sizeFactor&#39;</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&#39;denoised PCA&#39;</span>)</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
</div>
<div class="cell-output-display">
<p><img src="Lab5_clustering_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672" /></p>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="clustering" class="level2" data-number="9.2">
<h2 data-number="9.2"><span class="header-section-number">9.2</span> Clustering</h2>
<p>Clustering is an unsupervised learning procedure that is used in scRNA-seq data analysis to empirically define groups of cells with similar expression profiles. Its primary purpose is to summarize the data in a digestible format for human interpretation.</p>
<p>After annotation based on marker genes, the clusters can be treated as proxies for more abstract biological concepts such as cell types or states. Clustering is thus a critical step for extracting biological insights from scRNA-seq data.</p>
<section id="clustering-algorithms" class="level3" data-number="9.2.1">
<h3 data-number="9.2.1"><span class="header-section-number">9.2.1</span> Clustering algorithms</h3>
<p>Three main approaches can be used:</p>
<ol type="1">
<li>Hierarchical clustering</li>
<li>k-means clustering</li>
<li>Graph-based clustering</li>
</ol>
<p>Today, we will focus on graph-based clustering, as it is becoming the standard for scRNAseq: it is a flexible and scalable technique for clustering even the largest scRNA-seq datasets. We first build a graph where each node is a cell that is connected by edges to its nearest neighbors in the high-dimensional space. Edges are weighted based on the similarity between the cells involved, with higher weight given to cells that are more closely related.</p>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Compute a graph-based clustering of the PBMC dataset.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-13-contents" aria-controls="callout-13" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-13" class="callout-13-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb29"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>graph <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">buildSNNGraph</span>(</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>    sce, </span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">k =</span> <span class="dv">5</span>, </span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">use.dimred =</span> <span class="st">&#39;PCA&#39;</span></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>sce_clust <span class="ot">&lt;-</span> igraph<span class="sc">::</span><span class="fu">cluster_louvain</span>(graph)<span class="sc">$</span>membership</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>clusters_graph <span class="ot">&lt;-</span> <span class="fu">factor</span>(sce_clust)</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(sce<span class="sc">$</span>clusters_graph)</span></code></pre></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17 
502 359 682  40 129 393  61 193 349 363 376 218 190 151 127 197  10 </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>What are the main parameters to choose when building the graph? How do they impact the clustering?</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-15-contents" aria-controls="callout-15" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-15" class="callout-15-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb31"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>graph2 <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">buildSNNGraph</span>(</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>    sce, </span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">k =</span> <span class="dv">50</span>, </span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">use.dimred =</span> <span class="st">&#39;PCA&#39;</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>sce_clust2 <span class="ot">&lt;-</span> igraph<span class="sc">::</span><span class="fu">cluster_louvain</span>(graph2)<span class="sc">$</span>membership</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(sce_clust, sce_clust2)</span></code></pre></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>         sce_clust2
sce_clust   1   2   3   4   5   6   7   8
       1  502   0   0   0   0   0   0   0
       2    0 321  29   0   0   9   0   0
       3    0  23 659   0   0   0   0   0
       4    0   0   0  37   1   0   2   0
       5   23   0   0 106   0   0   0   0
       6    0   1   0   0 389   3   0   0
       7   58   0   1   0   2   0   0   0
       8    0 181  12   0   0   0   0   0
       9    0   1 348   0   0   0   0   0
       10 363   0   0   0   0   0   0   0
       11   0   1   0   0   0 374   1   0
       12   0   0   0   0 218   0   0   0
       13   0   0   0   0   1  22 167   0
       14   0   0   0   0   0 151   0   0
       15  14   0   0   1   0   0   0 112
       16   0 167   1   0   0  29   0   0
       17   0   5   2   0   3   0   0   0</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="dimensional-reduction-for-clustering-visualization" class="level3" data-number="9.2.2">
<h3 data-number="9.2.2"><span class="header-section-number">9.2.2</span> Dimensional reduction for clustering visualization</h3>
<p><code>PCA</code> is a powerful linear approach to compress large datasets into smaller dimensional spaces. However, it struggles at emphasizing the existence of clusters in complex datasets, when visualized in 2D.</p>
<p><code>scater</code> provides a handy way to perform more complex data embeddings:</p>
<pre><code>- tSNE
- UMAP
- Diffusion Map
- Multi-Dimensional Scaling (MDS)
- Non-negative Matrix Factorization (NMF)</code></pre>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Explore the different dimensional reduction algorithms, trying different hyperparameters combinations.</p>
<p>When you run these commands, pay attention to how long each command takes to run! While this run, check the <code>Help</code> page for each function (e.g. <code>?runTSNE</code>)</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-17-contents" aria-controls="callout-17" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-17" class="callout-17-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb34"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">reducedDims</span>(sce)</span></code></pre></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>List of length 1
names(1): PCA</code></pre>
</div>
<div class="sourceCode" id="cb36"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> scater<span class="sc">::</span><span class="fu">runTSNE</span>(sce, <span class="at">subset_row =</span> HVGs)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> scater<span class="sc">::</span><span class="fu">runUMAP</span>(sce, <span class="at">subset_row =</span> HVGs)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="fu">reducedDims</span>(sce)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of length 3
names(3): PCA TSNE UMAP</code></pre>
</div>
<div class="sourceCode" id="cb38"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">reducedDim</span>(sce, <span class="st">&#39;UMAP&#39;</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, ]</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        UMAP1    UMAP2
 [1,] 10.4762 -1.74452
 [2,]  9.9714 -1.70896
 [3,]  8.9500 -1.36885
 [4,] -6.3458 -3.38246
 [5,] -5.7974 -6.26426
 [6,] -6.6016 -1.30995
 [7,]  8.3734 11.37350
 [8,]  8.4162  2.35892
 [9,]  2.7606 14.29711
[10,]  8.8376  0.10778</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Use the <code>scater::plotReducedDim()</code> function to plot cells in each embedding. Comment.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-19-contents" aria-controls="callout-19" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-19" class="callout-19-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb40"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>p<span class="ot">&lt;-</span> scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(sce, <span class="st">&#39;PCA&#39;</span>, <span class="at">colour_by =</span> <span class="st">&#39;clusters_graph&#39;</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&#39;denoised PCA&#39;</span>) <span class="sc">+</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>    scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(sce, <span class="st">&#39;TSNE&#39;</span>, <span class="at">colour_by =</span> <span class="st">&#39;clusters_graph&#39;</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&#39;tSNE&#39;</span>) <span class="sc">+</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>    scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(sce, <span class="st">&#39;UMAP&#39;</span>, <span class="at">colour_by =</span> <span class="st">&#39;clusters_graph&#39;</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&#39;UMAP&#39;</span>)</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
</div>
<div class="cell-output-display">
<p><img src="Lab5_clustering_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672" /></p>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="bonus-for-the-pros-of-clustering-compare-different-clustering-approaches" class="level3" data-number="9.2.3">
<h3 data-number="9.2.3"><span class="header-section-number">9.2.3</span> <a href="#bonus">BONUS</a> For the pros of clustering… Compare different clustering approaches</h3>
<p>Leveraging the <code>bluster</code> package, different clustering approaches can be performed using a uniformed syntax, to compare their output.</p>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Using <code>clusterSweep()</code>, compare the effect of different <code>k</code> neighbor values when performing graph-based clustering.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-21-contents" aria-controls="callout-21" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-21" class="callout-21-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb41"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>clusters <span class="ot">&lt;-</span> bluster<span class="sc">::</span><span class="fu">clusterSweep</span>(</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">reducedDim</span>(sce, <span class="st">&#39;PCA&#39;</span>), </span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">BLUSPARAM =</span> bluster<span class="sc">::</span><span class="fu">SNNGraphParam</span>(),</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">k =</span> <span class="fu">c</span>(5L, 15L, 25L, 50L), </span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">cluster.fun =</span> <span class="fu">c</span>(<span class="st">&quot;louvain&quot;</span>)</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(clusters<span class="sc">$</span>clusters)</span></code></pre></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] &quot;k.5_cluster.fun.louvain&quot;  &quot;k.15_cluster.fun.louvain&quot; &quot;k.25_cluster.fun.louvain&quot; &quot;k.50_cluster.fun.louvain&quot;</code></pre>
</div>
<div class="sourceCode" id="cb43"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(clusters<span class="sc">$</span>clusters)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 4 columns
  k.5_cluster.fun.louvain k.15_cluster.fun.louvain k.25_cluster.fun.louvain k.50_cluster.fun.louvain
                 &lt;factor&gt;                 &lt;factor&gt;                 &lt;factor&gt;                 &lt;factor&gt;
1                       1                        1                        1                        1
2                       2                        1                        1                        1
3                       1                        1                        1                        1
4                       3                        2                        2                        2
5                       4                        3                        3                        3
6                       3                        2                        3                        3</code></pre>
</div>
<div class="sourceCode" id="cb45"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>clusters<span class="sc">$</span>parameters</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 4 rows and 2 columns
                                 k cluster.fun
                         &lt;integer&gt; &lt;character&gt;
k.5_cluster.fun.louvain          5     louvain
k.15_cluster.fun.louvain        15     louvain
k.25_cluster.fun.louvain        25     louvain
k.50_cluster.fun.louvain        50     louvain</code></pre>
</div>
<div class="sourceCode" id="cb47"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggraph)</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>    clustree<span class="sc">::</span><span class="fu">clustree</span>(</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>        clusters<span class="sc">$</span>clusters <span class="sc">%&gt;%</span> <span class="fu">setNames</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(.)) <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>(),</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">prefix =</span> <span class="st">&#39;X&#39;</span>,</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">edge_arrow=</span><span class="cn">FALSE</span></span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>    cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>        scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(sce, <span class="st">&#39;TSNE&#39;</span>, <span class="at">colour_by =</span> <span class="fu">I</span>(clusters<span class="sc">$</span>clusters[, <span class="st">&#39;k.5_cluster.fun.louvain&#39;</span>])) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&#39;k = 5&#39;</span>),</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>        scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(sce, <span class="st">&#39;TSNE&#39;</span>, <span class="at">colour_by =</span> <span class="fu">I</span>(clusters<span class="sc">$</span>clusters[, <span class="st">&#39;k.15_cluster.fun.louvain&#39;</span>])) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&#39;k = 15&#39;</span>),</span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>        scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(sce, <span class="st">&#39;TSNE&#39;</span>, <span class="at">colour_by =</span> <span class="fu">I</span>(clusters<span class="sc">$</span>clusters[, <span class="st">&#39;k.25_cluster.fun.louvain&#39;</span>])) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&#39;k = 25&#39;</span>),</span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a>        scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(sce, <span class="st">&#39;TSNE&#39;</span>, <span class="at">colour_by =</span> <span class="fu">I</span>(clusters<span class="sc">$</span>clusters[, <span class="st">&#39;k.50_cluster.fun.louvain&#39;</span>])) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&#39;k = 50&#39;</span>)</span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a>    ), </span>
<span id="cb47-14"><a href="#cb47-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">nrow =</span> <span class="dv">2</span>, </span>
<span id="cb47-15"><a href="#cb47-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">rel_heights =</span> <span class="fu">c</span>(<span class="fl">0.3</span>, <span class="fl">0.7</span>)</span>
<span id="cb47-16"><a href="#cb47-16" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb47-17"><a href="#cb47-17" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<div class="cell-output-display">
<p><img src="Lab5_clustering_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672" /></p>
</div>
<div class="sourceCode" id="cb48"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(clusters<span class="sc">$</span>clusters[, <span class="st">&#39;k.5_cluster.fun.louvain&#39;</span>])</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18 
473 393 368 627  40 129 376  61 188 404 382 217 190 141 126 197  10  18 </code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="cell-annotation" class="level2" data-number="9.3">
<h2 data-number="9.3"><span class="header-section-number">9.3</span> Cell annotation</h2>
<section id="find-marker-genes" class="level3" data-number="9.3.1">
<h3 data-number="9.3.1"><span class="header-section-number">9.3.1</span> Find marker genes</h3>
<p>To interpret clustering results, one needs to identify the genes that drive separation between clusters. These marker genes allow to assign biological meaning to each cluster based on their functional annotation. In the most obvious case, the marker genes for each cluster are <em>a priori</em> associated with particular cell types, allowing us to treat the clustering as a <em>proxy</em> for cell type identity.</p>
<p>A general strategy is to perform DE tests between pairs of clusters and then combine results into a single ranking of marker genes for each cluster.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb50"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">findMarkers</span>(sce, <span class="at">groups =</span> sce<span class="sc">$</span>clusters_graph)</span></code></pre></div>
</div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Find markers strongly overexpressed in each cluster. Check <code>?scran::findMarkers</code> to find the right options to use.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-23-contents" aria-controls="callout-23" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-23" class="callout-23-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb51"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> scran<span class="sc">::</span><span class="fu">findMarkers</span>(</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>    sce, </span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">groups =</span> sce<span class="sc">$</span>clusters_graph, </span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">direction =</span> <span class="st">&quot;up&quot;</span>, </span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">lfc =</span> <span class="dv">1</span></span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(markers[[<span class="dv">1</span>]])</span></code></pre></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 20 columns
                    Top      p.value          FDR summary.logFC   logFC.2   logFC.3   logFC.4   logFC.5   logFC.6   logFC.7   logFC.8   logFC.9  logFC.10  logFC.11  logFC.12  logFC.13  logFC.14
              &lt;integer&gt;    &lt;numeric&gt;    &lt;numeric&gt;     &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
S100A9                1  0.00000e+00  0.00000e+00       5.41682   5.43237   5.41682   5.38506  3.527516   5.40582  1.300493   5.30313   5.29337  1.367411   5.51013   5.25562   5.57793   5.39685
S100A12               1 3.00254e-186 7.78211e-183       3.16760   3.14991   3.15614   3.06390  2.752834   3.16872  1.321253   3.13691   3.11275  1.428806   3.15624   3.14479   3.16760   3.13757
S100A8                1  0.00000e+00  0.00000e+00       5.35071   5.40804   5.35071   5.24867  4.002828   5.36811  1.429670   5.33249   5.42634  1.615361   5.44104   5.37566   5.57562   5.42914
MNDA                  1 6.44450e-174 1.35713e-170       2.69302   2.62331   2.65953   2.68560  1.039333   2.63823  0.869333   2.62066   2.66382  0.469773   2.68082   2.63842   2.69302   2.67460
LYZ                   1  0.00000e+00  0.00000e+00       5.27217   5.25980   5.27217   5.32183  0.755057   5.22486  0.992756   5.33170   5.26958  0.312196   5.30534   5.30214   5.49214   5.27623
RP11-1143G9.4         1 6.20347e-200 1.74183e-196       3.22726   3.16209   3.16589   3.04120  1.380130   3.16979  0.939639   3.18494   3.16750  0.865393   3.18003   3.17216   3.18095   3.22726
               logFC.15  logFC.16  logFC.17
              &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
S100A9          4.11810   5.40420   5.33767
S100A12         3.05345   3.10968   3.06099
S100A8          4.62013   5.46970   5.41967
MNDA            1.17765   2.66604   2.70577
LYZ             3.03608   5.25555   5.19250
RP11-1143G9.4   2.72497   3.22522   3.21302</code></pre>
</div>
<div class="sourceCode" id="cb53"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">&lt;-</span> <span class="fu">lapply</span>(markers, <span class="cf">function</span>(df) {</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>(df[df<span class="sc">$</span>Top <span class="sc">&lt;=</span> <span class="dv">5</span>,])</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>})</span></code></pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Plot average expression of the first marker of the first cluster in UMAP.</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-25-contents" aria-controls="callout-25" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-25" class="callout-25-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb54"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(sce, <span class="st">&#39;TSNE&#39;</span>, <span class="at">colour_by =</span> markers[[<span class="dv">2</span>]][[<span class="dv">1</span>]])</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
</div>
<div class="cell-output-display">
<p><img src="Lab5_clustering_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672" /></p>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="automated-cell-annotation" class="level3" data-number="9.3.2">
<h3 data-number="9.3.2"><span class="header-section-number">9.3.2</span> Automated cell annotation</h3>
<p>Many cell type reference databases are available over the Internet. Today, we will use a reference constructed from <code>Blueprint</code> and <code>ENCODE</code> data (<code>Martens and Stunnenberg 2013</code>; <code>The ENCODE Project Consortium 2012</code>). This reference is available as a <code>SummarizedExperiment</code> containing log-normalized gene expression for manually annotated samples.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb55"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>ref <span class="ot">&lt;-</span> celldex<span class="sc">::</span><span class="fu">BlueprintEncodeData</span>()</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>prediction_types <span class="ot">&lt;-</span> SingleR<span class="sc">::</span><span class="fu">SingleR</span>(</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">test =</span> sce, </span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">ref =</span> ref, </span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels =</span> ref<span class="sc">$</span>label.main</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>annotation <span class="ot">&lt;-</span> prediction_types<span class="sc">$</span>labels</span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(sce<span class="sc">$</span>annotation)</span></code></pre></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>
     B-cells CD4+ T-cells CD8+ T-cells           DC          HSC    Monocytes     NK cells 
         616          881         1390            1           14         1201          237 </code></pre>
</div>
<div class="sourceCode" id="cb57"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(sce<span class="sc">$</span>annotation, sce<span class="sc">$</span>clusters_graph)</span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              
                 1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  B-cells        0   0   0  14   0 383   0   0   0   0   0 218   0   0   0   0   1
  CD4+ T-cells   0 158 490   0   0   0   1  96  47   0   0   0   0   0   0  89   0
  CD8+ T-cells   0 201 192   0   0   9   0  97 302   0 342   0   3 136   0 108   0
  DC             0   0   0   0   1   0   0   0   0   0   0   0   0   0   0   0   0
  HSC            0   0   0   5   0   0   0   0   0   0   0   0   0   0   0   0   9
  Monocytes    502   0   0  21 128   0  60   0   0 363   0   0   0   0 127   0   0
  NK cells       0   0   0   0   0   1   0   0   0   0  34   0 187  15   0   0   0</code></pre>
</div>
</div>
<div class="icon callout callout-style-default callout-question no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Using <code>scater</code> and <code>SingleR</code> utilities, visually compare the annotation scores for cells in each cluster.</p>
<p>Did the automated annotation work robuslty? How does it compare to our clustering? Is automated annotation as sensitive as graph-based clustering?</p>
<div class="icon callout callout-style-default callout-answer no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-27-contents" aria-controls="callout-27" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class='callout-icon no-icon'></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class='callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end'><i class='callout-toggle'></i></div>
</div>
<div id="callout-27" class="callout-27-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb59"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> SingleR<span class="sc">::</span><span class="fu">plotScoreHeatmap</span>(prediction_types)</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
</div>
<div class="cell-output-display">
<p><img src="Lab5_clustering_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672" /></p>
</div>
<div class="sourceCode" id="cb60"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> scater<span class="sc">::</span><span class="fu">plotReducedDim</span>(sce, <span class="st">&#39;TSNE&#39;</span>, <span class="at">colour_by =</span> <span class="st">&#39;annotation&#39;</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">&#39;Automated annotation&#39;</span>)</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>p</span></code></pre></div>
<div class="cell-output-display">
<p><img src="Lab5_clustering_files/figure-html/unnamed-chunk-17-2.png" class="img-fluid" width="672" /></p>
</div>
<div class="sourceCode" id="cb61"><pre class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> pheatmap<span class="sc">::</span><span class="fu">pheatmap</span>(</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">log2</span>(<span class="fu">table</span>(<span class="at">Annotation =</span> sce<span class="sc">$</span>annotation, <span class="at">Cluster =</span> sce<span class="sc">$</span>clusters_graph)<span class="sc">+</span><span class="dv">10</span>), </span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">color =</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">&quot;white&quot;</span>, <span class="st">&quot;darkred&quot;</span>))(<span class="dv">101</span>)</span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<div class="cell-output-display">
<p><img src="Lab5_clustering_files/figure-html/unnamed-chunk-17-3.png" class="img-fluid" width="672" /></p>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="bonus" class="level2" data-number="9.4">
<h2 data-number="9.4"><span class="header-section-number">9.4</span> Bonus</h2>
<p>Try to fill in the analysis template in <code>bin/prepare_Ernst.R</code> to execute the different processing/analysis steps we covered in the previous exercises and this one. If you prefer using <code>Seurat</code>, don’t hesitate to modify the base template!</p>
</section>
<section id="acknowledgements" class="level2" data-number="9.5">
<h2 data-number="9.5"><span class="header-section-number">9.5</span> Acknowledgements</h2>
<p>This exercise was adapted from Chapts. 7-12 of <a href="https://bioconductor.org/books/release/OSCA/">Orchestrating Single-Cell Analysis with Bioconductor</a>.</p>
</section>
<section id="session-info" class="level2" data-number="9.6">
<h2 data-number="9.6"><span class="header-section-number">9.6</span> Session info</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.4.1 (2024-06-14)
 os       macOS Sonoma 14.6.1
 system   aarch64, darwin20
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       Europe/Paris
 date     2024-11-03
 pandoc   2.19.2 @ /opt/homebrew/bin/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 package              * version  date (UTC) lib source
 abind                * 1.4-8    2024-09-12 [1] CRAN (R 4.4.1)
 alabaster.base         1.4.2    2024-06-23 [1] Bioconduc~
 alabaster.matrix       1.4.2    2024-06-23 [1] Bioconduc~
 alabaster.ranges       1.4.2    2024-06-23 [1] Bioconduc~
 alabaster.schemas      1.4.0    2024-04-30 [1] Bioconduc~
 alabaster.se           1.4.1    2024-05-30 [1] Bioconduc~
 AnnotationDbi          1.66.0   2024-06-16 [1] Bioconduc~
 AnnotationHub          3.12.0   2024-04-30 [1] Bioconduc~
 backports              1.5.0    2024-05-23 [1] CRAN (R 4.4.0)
 beachmat               2.20.0   2024-05-06 [1] Bioconduc~
 beeswarm               0.4.0    2021-06-01 [1] CRAN (R 4.4.0)
 Biobase              * 2.64.0   2024-04-30 [1] Bioconduc~
 BiocFileCache          2.12.0   2024-04-30 [1] Bioconduc~
 BiocGenerics         * 0.50.0   2024-04-30 [1] Bioconduc~
 BiocManager            1.30.25  2024-08-28 [1] CRAN (R 4.4.1)
 BiocNeighbors          1.22.0   2024-04-30 [1] Bioconduc~
 BiocParallel           1.38.0   2024-04-30 [1] Bioconduc~
 BiocSingular           1.20.0   2024-04-30 [1] Bioconduc~
 BiocVersion            3.19.1   2024-04-22 [1] Bioconduc~
 Biostrings             2.72.1   2024-06-02 [1] Bioconduc~
 bit                    4.5.0    2024-09-20 [1] CRAN (R 4.4.1)
 bit64                  4.5.2    2024-09-22 [1] CRAN (R 4.4.1)
 blob                   1.2.4    2023-03-17 [1] CRAN (R 4.4.0)
 bluster                1.14.0   2024-04-30 [1] Bioconduc~
 cachem                 1.1.0    2024-05-16 [1] CRAN (R 4.4.0)
 celldex                1.14.0   2024-05-02 [1] Bioconduc~
 checkmate              2.3.2    2024-07-29 [1] CRAN (R 4.4.0)
 cli                    3.6.3    2024-06-21 [1] CRAN (R 4.4.0)
 cluster                2.1.6    2023-12-01 [1] CRAN (R 4.4.1)
 clustree               0.5.1    2023-11-05 [1] CRAN (R 4.4.0)
 codetools              0.2-20   2024-03-31 [1] CRAN (R 4.4.1)
 colorspace             2.1-1    2024-07-26 [1] CRAN (R 4.4.0)
 cowplot                1.1.3    2024-01-22 [1] CRAN (R 4.4.0)
 crayon                 1.5.3    2024-06-20 [1] CRAN (R 4.4.0)
 curl                   5.2.3    2024-09-20 [1] CRAN (R 4.4.1)
 DBI                    1.2.3    2024-06-02 [1] CRAN (R 4.4.0)
 dbplyr                 2.5.0    2024-03-19 [1] CRAN (R 4.4.0)
 DelayedArray         * 0.30.1   2024-05-30 [1] Bioconduc~
 DelayedMatrixStats     1.26.0   2024-04-30 [1] Bioconduc~
 devtools               2.4.5    2022-10-11 [1] CRAN (R 4.4.0)
 digest                 0.6.37   2024-08-19 [1] CRAN (R 4.4.1)
 dplyr                * 1.1.4    2023-11-17 [1] CRAN (R 4.4.0)
 dqrng                  0.4.1    2024-05-28 [1] CRAN (R 4.4.0)
 edgeR                  4.2.2    2024-10-13 [1] Bioconduc~
 ellipsis               0.3.2    2021-04-29 [1] CRAN (R 4.4.0)
 evaluate               1.0.1    2024-10-10 [1] CRAN (R 4.4.1)
 ExperimentHub          2.12.0   2024-04-30 [1] Bioconduc~
 fansi                  1.0.6    2023-12-08 [1] CRAN (R 4.4.0)
 farver                 2.1.2    2024-05-13 [1] CRAN (R 4.4.0)
 fastmap                1.2.0    2024-05-15 [1] CRAN (R 4.4.0)
 filelock               1.0.3    2023-12-11 [1] CRAN (R 4.4.0)
 forcats              * 1.0.0    2023-01-29 [1] CRAN (R 4.4.0)
 fs                     1.6.4    2024-04-25 [1] CRAN (R 4.4.0)
 generics               0.1.3    2022-07-05 [1] CRAN (R 4.4.0)
 GenomeInfoDb         * 1.40.1   2024-06-16 [1] Bioconduc~
 GenomeInfoDbData       1.2.12   2024-10-29 [1] Bioconductor
 GenomicRanges        * 1.56.2   2024-10-09 [1] Bioconduc~
 ggbeeswarm             0.7.2    2023-04-29 [1] CRAN (R 4.4.0)
 ggforce                0.4.2    2024-02-19 [1] CRAN (R 4.4.0)
 ggplot2              * 3.5.1    2024-04-23 [1] CRAN (R 4.4.0)
 ggraph               * 2.2.1    2024-03-07 [1] CRAN (R 4.4.0)
 ggrepel                0.9.6    2024-09-07 [1] CRAN (R 4.4.1)
 glue                   1.8.0    2024-09-30 [1] CRAN (R 4.4.1)
 graphlayouts           1.2.0    2024-09-24 [1] CRAN (R 4.4.1)
 gridExtra              2.3      2017-09-09 [1] CRAN (R 4.4.0)
 gtable                 0.3.6    2024-10-25 [1] CRAN (R 4.4.1)
 gypsum                 1.0.1    2024-05-30 [1] Bioconduc~
 HDF5Array            * 1.32.1   2024-08-11 [1] Bioconduc~
 hms                    1.1.3    2023-03-21 [1] CRAN (R 4.4.0)
 htmltools              0.5.8.1  2024-04-04 [1] CRAN (R 4.4.0)
 htmlwidgets            1.6.4    2023-12-06 [1] CRAN (R 4.4.0)
 httpuv                 1.6.15   2024-03-26 [1] CRAN (R 4.4.0)
 httr                   1.4.7    2023-08-15 [1] CRAN (R 4.4.0)
 httr2                  1.0.5    2024-09-26 [1] CRAN (R 4.4.1)
 igraph                 2.1.1    2024-10-19 [1] CRAN (R 4.4.1)
 IRanges              * 2.38.1   2024-07-03 [1] Bioconduc~
 irlba                  2.3.5.1  2022-10-03 [1] CRAN (R 4.4.0)
 jsonlite               1.8.9    2024-09-20 [1] CRAN (R 4.4.1)
 KEGGREST               1.44.1   2024-06-19 [1] Bioconduc~
 knitr                  1.48     2024-07-07 [1] CRAN (R 4.4.0)
 labeling               0.4.3    2023-08-29 [1] CRAN (R 4.4.0)
 later                  1.3.2    2023-12-06 [1] CRAN (R 4.4.0)
 lattice                0.22-6   2024-03-20 [1] CRAN (R 4.4.1)
 lifecycle              1.0.4    2023-11-07 [1] CRAN (R 4.4.0)
 limma                  3.60.6   2024-10-02 [1] Bioconduc~
 locfit                 1.5-9.10 2024-06-24 [1] CRAN (R 4.4.0)
 lubridate            * 1.9.3    2023-09-27 [1] CRAN (R 4.4.0)
 magrittr               2.0.3    2022-03-30 [1] CRAN (R 4.4.0)
 MASS                   7.3-60.2 2024-04-26 [1] CRAN (R 4.4.1)
 Matrix               * 1.7-1    2024-10-18 [1] CRAN (R 4.4.1)
 MatrixGenerics       * 1.16.0   2024-04-30 [1] Bioconduc~
 matrixStats          * 1.4.1    2024-09-08 [1] CRAN (R 4.4.1)
 memoise                2.0.1    2021-11-26 [1] CRAN (R 4.4.0)
 metapod                1.12.0   2024-04-30 [1] Bioconduc~
 mime                   0.12     2021-09-28 [1] CRAN (R 4.4.0)
 miniUI                 0.1.1.1  2018-05-18 [1] CRAN (R 4.4.0)
 munsell                0.5.1    2024-04-01 [1] CRAN (R 4.4.0)
 patchwork            * 1.3.0    2024-09-16 [1] CRAN (R 4.4.1)
 pheatmap               1.0.12   2019-01-04 [1] CRAN (R 4.4.0)
 pillar                 1.9.0    2023-03-22 [1] CRAN (R 4.4.0)
 pkgbuild               1.4.5    2024-10-28 [1] CRAN (R 4.4.1)
 pkgconfig              2.0.3    2019-09-22 [1] CRAN (R 4.4.0)
 pkgload                1.4.0    2024-06-28 [1] CRAN (R 4.4.0)
 png                    0.1-8    2022-11-29 [1] CRAN (R 4.4.0)
 polyclip               1.10-7   2024-07-23 [1] CRAN (R 4.4.0)
 profvis                0.4.0    2024-09-20 [1] CRAN (R 4.4.1)
 promises               1.3.0    2024-04-05 [1] CRAN (R 4.4.0)
 purrr                * 1.0.2    2023-08-10 [1] CRAN (R 4.4.0)
 R6                     2.5.1    2021-08-19 [1] CRAN (R 4.4.0)
 rappdirs               0.3.3    2021-01-31 [1] CRAN (R 4.4.0)
 RColorBrewer           1.1-3    2022-04-03 [1] CRAN (R 4.4.0)
 Rcpp                   1.0.13   2024-07-17 [1] CRAN (R 4.4.0)
 RcppAnnoy              0.0.22   2024-01-23 [1] CRAN (R 4.4.0)
 readr                * 2.1.5    2024-01-10 [1] CRAN (R 4.4.0)
 remotes                2.5.0    2024-03-17 [1] CRAN (R 4.4.0)
 rhdf5                * 2.48.0   2024-04-30 [1] Bioconduc~
 rhdf5filters           1.16.0   2024-04-30 [1] Bioconduc~
 Rhdf5lib               1.26.0   2024-04-30 [1] Bioconduc~
 rlang                  1.1.4    2024-06-04 [1] CRAN (R 4.4.0)
 rmarkdown              2.28     2024-08-17 [1] CRAN (R 4.4.0)
 RSQLite                2.3.7    2024-05-27 [1] CRAN (R 4.4.0)
 rsvd                   1.0.5    2021-04-16 [1] CRAN (R 4.4.0)
 Rtsne                  0.17     2023-12-07 [1] CRAN (R 4.4.0)
 S4Arrays             * 1.4.1    2024-05-30 [1] Bioconduc~
 S4Vectors            * 0.42.1   2024-07-03 [1] Bioconduc~
 ScaledMatrix           1.12.0   2024-04-30 [1] Bioconduc~
 scales                 1.3.0    2023-11-28 [1] CRAN (R 4.4.0)
 scater                 1.32.1   2024-07-21 [1] Bioconduc~
 scran                  1.32.0   2024-04-30 [1] Bioconduc~
 scuttle                1.14.0   2024-04-30 [1] Bioconduc~
 sessioninfo            1.2.2    2021-12-06 [1] CRAN (R 4.4.0)
 shiny                  1.9.1    2024-08-01 [1] CRAN (R 4.4.0)
 SingleCellExperiment * 1.26.0   2024-04-30 [1] Bioconduc~
 SingleR                2.6.0    2024-04-30 [1] Bioconduc~
 SparseArray          * 1.4.8    2024-05-30 [1] Bioconduc~
 sparseMatrixStats      1.16.0   2024-04-30 [1] Bioconduc~
 statmod                1.5.0    2023-01-06 [1] CRAN (R 4.4.0)
 stringi                1.8.4    2024-05-06 [1] CRAN (R 4.4.0)
 stringr              * 1.5.1    2023-11-14 [1] CRAN (R 4.4.0)
 SummarizedExperiment * 1.34.0   2024-04-30 [1] Bioconduc~
 TENxPBMCData         * 1.22.0   2024-05-02 [1] Bioconduc~
 tibble               * 3.2.1    2023-03-20 [1] CRAN (R 4.4.0)
 tidygraph              1.3.1    2024-01-30 [1] CRAN (R 4.4.0)
 tidyr                * 1.3.1    2024-01-24 [1] CRAN (R 4.4.0)
 tidyselect             1.2.1    2024-03-11 [1] CRAN (R 4.4.0)
 tidyverse            * 2.0.0    2023-02-22 [1] CRAN (R 4.4.0)
 timechange             0.3.0    2024-01-18 [1] CRAN (R 4.4.0)
 tweenr                 2.0.3    2024-02-26 [1] CRAN (R 4.4.0)
 tzdb                   0.4.0    2023-05-12 [1] CRAN (R 4.4.0)
 UCSC.utils             1.0.0    2024-05-06 [1] Bioconduc~
 urlchecker             1.0.1    2021-11-30 [1] CRAN (R 4.4.0)
 usethis                3.0.0    2024-07-29 [1] CRAN (R 4.4.0)
 utf8                   1.2.4    2023-10-22 [1] CRAN (R 4.4.0)
 uwot                   0.2.2    2024-04-21 [1] CRAN (R 4.4.0)
 vctrs                  0.6.5    2023-12-01 [1] CRAN (R 4.4.0)
 vipor                  0.4.7    2023-12-18 [1] CRAN (R 4.4.0)
 viridis                0.6.5    2024-01-29 [1] CRAN (R 4.4.0)
 viridisLite            0.4.2    2023-05-02 [1] CRAN (R 4.4.0)
 withr                  3.0.2    2024-10-28 [1] CRAN (R 4.4.1)
 xfun                   0.48     2024-10-03 [1] CRAN (R 4.4.1)
 xtable                 1.8-4    2019-04-21 [1] CRAN (R 4.4.0)
 XVector                0.44.0   2024-04-30 [1] Bioconduc~
 yaml                   2.3.10   2024-07-26 [1] CRAN (R 4.4.0)
 zlibbioc               1.50.0   2024-04-30 [1] Bioconduc~

 [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library

──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>
<div id="quarto-navigation-envelope" class="hidden">
<p><span class="hidden" data-render-id="quarto-int-sidebar-title">Single-cell RNAseq analysis with R/Bioconductor</span> <span class="hidden" data-render-id="quarto-int-navbar-title">Single-cell RNAseq analysis with R/Bioconductor</span> <span class="hidden" data-render-id="quarto-int-next"><span class="chapter-number">10</span>  <span class="chapter-title">Lecture 5 - Data integration and batch effect correction</span></span> <span class="hidden" data-render-id="quarto-int-prev"><span class="chapter-number">8</span>  <span class="chapter-title">Lecture 4 - Identifying cell populations</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:/index.html">Welcome</span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/program.html">Program</span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/Rstudio.html">RStudio</span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/prerequisites.html">Prerequisites</span> <span class="hidden" data-render-id="quarto-int-sidebar:quarto-sidebar-section-1">Day 1</span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/day1/Lecture1_introduction.html"><span class="chapter-number">1</span>  <span class="chapter-title">Lecture 1 - Introduction to scRNAseq analysis</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/day1/Lecture2_Bcl2matrix.html"><span class="chapter-number">2</span>  <span class="chapter-title">Lecture 2 - From sequencing reads to expression matrices</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/day1/Lab1_Intro.html"><span class="chapter-number">3</span>  <span class="chapter-title">Lab 1: Familiarizing yourself with the course AWS instance</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/day1/Lab2_processingreads.html"><span class="chapter-number">4</span>  <span class="chapter-title">Lab 2: From .bcl to count matrix</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:quarto-sidebar-section-2">Day 2</span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/day2/Lecture3_qualitycontrol.html"><span class="chapter-number">5</span>  <span class="chapter-title">Lecture 3 - Quality control for scRNA-Seq data</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/day2/Lab3_Rbioc.html"><span class="chapter-number">6</span>  <span class="chapter-title">Lab 3: Introduction to R/Bioconductor</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/day2/Lab4_data_wrangling_scRNAseq.html"><span class="chapter-number">7</span>  <span class="chapter-title">Lab 4 - Single-cell RNA-seq data wrangling</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:quarto-sidebar-section-3">Day 3</span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/day3/Lecture4_clustering.html"><span class="chapter-number">8</span>  <span class="chapter-title">Lecture 4 - Identifying cell populations</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/day3/Lab5_clustering.html"><span class="chapter-number">9</span>  <span class="chapter-title">Lab 5: Dimension reduction, clustering and annotation</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/day3/Lecture5_batchcorrection.html"><span class="chapter-number">10</span>  <span class="chapter-title">Lecture 5 - Data integration and batch effect correction</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/day3/Lab6_batch_correction.html"><span class="chapter-number">11</span>  <span class="chapter-title">Lab 6: Batch correction</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:quarto-sidebar-section-4">Day 4</span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/day4/Lecture6_ATAC.html"><span class="chapter-number">12</span>  <span class="chapter-title">Lecture 6 - Advances in single-cell genomics: the epigenome</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/day4/Lab7_atac-seq.html"><span class="chapter-number">13</span>  <span class="chapter-title">Lab 7: Single-cell ATAC-seq analysis workflow</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/day4/Lecture7_pseudotime.html"><span class="chapter-number">14</span>  <span class="chapter-title">Lecture 7 - Trajectories and pseudotimes</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/day4/Lab8_pseudotime.html"><span class="chapter-number">15</span>  <span class="chapter-title">Lab 8: Pseudotime analyses</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:quarto-sidebar-section-5">Day 5</span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/day5/Lecture8_spatial-transcriptomics.html"><span class="chapter-number">16</span>  <span class="chapter-title">Lecture 8 - Advances in single-cell genomics: spatial transcriptomics</span></span> <span class="hidden" data-render-id="quarto-int-sidebar:/content/extra.html">Extra resources</span> <span class="hidden" data-render-id="footer-left">Single-cell RNAseq analysis with R/Bioconductor |<br />
J. Serizay, O. Ashenberg, F. Almeida-Silva</span> <span class="hidden" data-render-id="footer-right">This book was built with <a href="https://quarto.org/">Quarto</a>.</span> <span class="hidden" data-render-id="quarto-breadcrumbs-223e2965264ed5a7335008666fb8bf98">Day 3</span> <span class="hidden" data-render-id="quarto-breadcrumbs-940e2f3a7e083b30ba07ccf418401472"><span class="chapter-number">9</span>  <span class="chapter-title">Lab 5: Dimension reduction, clustering and annotation</span></span></p>
</div>
<div id="quarto-meta-markdown" class="hidden">
<p><span class="hidden" data-render-id="quarto-metatitle">Single-cell RNAseq analysis with R/Bioconductor - <span class="chapter-number">9</span>  <span class="chapter-title">Lab 5: Dimension reduction, clustering and annotation</span></span> <span class="hidden" data-render-id="quarto-twittercardtitle">Single-cell RNAseq analysis with R/Bioconductor - <span class="chapter-number">9</span>  <span class="chapter-title">Lab 5: Dimension reduction, clustering and annotation</span></span> <span class="hidden" data-render-id="quarto-ogcardtitle">Single-cell RNAseq analysis with R/Bioconductor - <span class="chapter-number">9</span>  <span class="chapter-title">Lab 5: Dimension reduction, clustering and annotation</span></span> <span class="hidden" data-render-id="quarto-metasitename">Single-cell RNAseq analysis with R/Bioconductor</span> <span class="hidden" data-render-id="quarto-twittercarddesc"></span> <span class="hidden" data-render-id="quarto-ogcardddesc"></span></p>
</div>
</section>

</main> <!-- /main -->
<script id = "quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a  href="/content/day3/Lecture4_clustering.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class='chapter-number'>8</span>  <span class='chapter-title'>Lecture 4 - Identifying cell populations</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a  href="/content/day3/Lecture5_batchcorrection.html" class="pagination-link">
        <span class="nav-page-text"><span class='chapter-number'>10</span>  <span class='chapter-title'>Lecture 5 - Data integration and batch effect correction</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      <div class='footer-contents'>Single-cell RNAseq analysis with R/Bioconductor |  
J. Serizay, O. Ashenberg, F. Almeida-Silva
</div>  
    </div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      <div class='footer-contents'>This book was built with <a href="https://quarto.org/">Quarto</a>.
</div>  
    </div>
  </div>
</footer>

</body>

</html>