<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.55">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="J. Serizay, O. Ashenberg &amp; F. Almeida-Silva">
<title>13&nbsp; Lab 7: Single-cell ATAC-seq analysis workflow â€“ Single-cell RNAseq analysis with R/Bioconductor</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../content/day4/Lecture7_pseudotime.html" rel="next">
<link href="../../content/day4/Lecture6_ATAC.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../content/day4/Lecture6_ATAC.html">Day 4</a></li><li class="breadcrumb-item"><a href="../../content/day4/Lab7_atac-seq.html"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Lab 7: Single-cell ATAC-seq analysis workflow</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../../">Single-cell RNAseq analysis with R/Bioconductor</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/js2264/scRNAseq_Physalia_2025/" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/program.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Program</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/Rstudio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">RStudio</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/prerequisites.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Prerequisites</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Day 1</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lecture1_introduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Lecture 1 - Introduction to scRNAseq analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lecture2_Bcl2matrix.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Lecture 2 - From sequencing reads to expression matrices</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lab1_Intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Lab 1: Familiarizing yourself with the course AWS instance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day1/Lab2_processingreads.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Lab 2: From .bcl to count matrix</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Day 2</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day2/Lecture3_qualitycontrol.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Lecture 3 - Quality control for scRNA-Seq data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day2/Lab3_Rbioc.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Lab 3: Introduction to R/Bioconductor</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day2/Lab4_data_wrangling_scRNAseq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Lab 4 - Single-cell RNA-seq data wrangling</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Day 3</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lecture4_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Lecture 4 - Identifying cell populations</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lab5_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Lab 5: Dimension reduction, clustering and annotation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lecture5_batchcorrection.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Lecture 5 - Data integration and batch effect correction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day3/Lab6_batch_correction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Lab 6: Batch correction</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Day 4</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lecture6_ATAC.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Lecture 6 - Advances in single-cell genomics: the epigenome</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lab7_atac-seq.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Lab 7: Single-cell ATAC-seq analysis workflow</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lecture7_pseudotime.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Lecture 7 - Trajectories and pseudotimes</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day4/Lab8_pseudotime.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Lab 8: Pseudotime analyses</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Day 5</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/day5/Lecture8_spatial-transcriptomics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Lecture 8 - Advances in single-cell genomics: spatial transcriptomics</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../content/extra.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Extra resources</span></a>
  </div>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#process-human-pbmc-dataset" id="toc-process-human-pbmc-dataset" class="nav-link active" data-scroll-target="#process-human-pbmc-dataset"><span class="header-section-number">13.1</span> Process human PBMC dataset</a>
  <ul class="collapse">
<li><a href="#import-data" id="toc-import-data" class="nav-link" data-scroll-target="#import-data"><span class="header-section-number">13.1.1</span> Import data</a></li>
  <li><a href="#create-a-seurat-object" id="toc-create-a-seurat-object" class="nav-link" data-scroll-target="#create-a-seurat-object"><span class="header-section-number">13.1.2</span> Create a <code>Seurat</code> object</a></li>
  <li><a href="#check-qcs" id="toc-check-qcs" class="nav-link" data-scroll-target="#check-qcs"><span class="header-section-number">13.1.3</span> Check QCs</a></li>
  <li><a href="#filter-cells-and-features" id="toc-filter-cells-and-features" class="nav-link" data-scroll-target="#filter-cells-and-features"><span class="header-section-number">13.1.4</span> Filter cells and features</a></li>
  <li><a href="#dimensionality-reduction-and-clustering" id="toc-dimensionality-reduction-and-clustering" class="nav-link" data-scroll-target="#dimensionality-reduction-and-clustering"><span class="header-section-number">13.1.5</span> Dimensionality reduction and clustering</a></li>
  </ul>
</li>
  <li><a href="#compute-gene-activity-scores" id="toc-compute-gene-activity-scores" class="nav-link" data-scroll-target="#compute-gene-activity-scores"><span class="header-section-number">13.2</span> Compute gene activity scores</a></li>
  <li><a href="#find-peaks-differentially-accessible-across-clusters" id="toc-find-peaks-differentially-accessible-across-clusters" class="nav-link" data-scroll-target="#find-peaks-differentially-accessible-across-clusters"><span class="header-section-number">13.3</span> Find peaks differentially accessible across clusters</a></li>
  <li><a href="#session-info" id="toc-session-info" class="nav-link" data-scroll-target="#session-info"><span class="header-section-number">13.4</span> Session info</a></li>
  </ul><div class="toc-actions"><ul><li><a href="https://github.com/js2264/scRNAseq_Physalia_2025/edit/main/content/day4/Lab7_atac-seq.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/js2264/scRNAseq_Physalia_2025/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../content/day4/Lecture6_ATAC.html">Day 4</a></li><li class="breadcrumb-item"><a href="../../content/day4/Lab7_atac-seq.html"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Lab 7: Single-cell ATAC-seq analysis workflow</span></a></li></ol></nav><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Lab 7: Single-cell ATAC-seq analysis workflow</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Notes
</div>
</div>
<div class="callout-body-container callout-body">
<p>The estimated time for this lab is around 1h10.</p>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Aims
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Import cells from human PBMC scATACseq dataset</li>
<li>Perform scATACseq quality controls and checks</li>
<li>Filter, normalize and plot PBMC scATACseq dataset</li>
<li>Compute gene activity scores, check known markers and annotate scATAC clusters</li>
<li>Perform differential peak accessibility analysis</li>
</ul>
</div>
</div>
<p>ATAC-seq data may be obtained in isolation using a single-cell ATAC-seq protocol (e.g.&nbsp;<a href="https://support.10xgenomics.com/single-cell-atac">10X scATACseq</a>) or in combination with gene expression data using a single-cell multiome protocole (e.g.&nbsp;<a href="https://www.10xgenomics.com/products/single-cell-multiome-atac-plus-gene-expression">10X multiome</a> and <a href="https://pubmed.ncbi.nlm.nih.gov/33098772/">SHARE-seq</a>).</p>
<p>Several packages are currently avaialble to process scATAC-seq data in R. These include <a href="https://satijalab.org/signac/index.html">Signac</a> and <a href="https://www.archrproject.com/bookdown/index.html#section">ArchR</a>. This lab will closely follow the <a href="https://satijalab.org/signac/articles/pbmc_vignette.html">processing steps</a> outlined in <code>Signac</code>, which interfaces well with <code>Seurat</code> for single-cell analysis.</p>
<p>In this lab, we will process a PBMC single-cell ATAC-seq (scATAC-seq) dataset and perform preliminary analysis to assess quality of these data. The data for this lab comes from <a href="https://www.10xgenomics.com/resources/datasets/10-k-peripheral-blood-mononuclear-cells-pbm-cs-from-a-healthy-donor-1-standard-1-0-1">10X Genomics</a>. The dataset contains roughly ~ 10,000 cells.</p>
<section id="process-human-pbmc-dataset" class="level2" data-number="13.1"><h2 data-number="13.1" class="anchored" data-anchor-id="process-human-pbmc-dataset">
<span class="header-section-number">13.1</span> Process human PBMC dataset</h2>
<section id="import-data" class="level3" data-number="13.1.1"><h3 data-number="13.1.1" class="anchored" data-anchor-id="import-data">
<span class="header-section-number">13.1.1</span> Import data</h3>
<p>Pre-downloaded data files are available in the <code>~/Share/scATAC/</code> folder. Notice how the count matrix is in a <code>.h5</code> format. We have already encountered this format in <a href="../../content/day2/Lab3_Rbioc.html">Lab3</a>. Back then, we imported it with <code><a href="https://rdrr.io/pkg/DropletUtils/man/read10xCounts.html">DropletUtils::read10xCounts</a></code>.</p>
<div class="icon callout callout-style-default callout-question callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Does this function work here?</p>
<div class="icon callout callout-style-default callout-answer callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">scatac</span> <span class="op">&lt;-</span> <span class="fu">DropletUtils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DropletUtils/man/read10xCounts.html">read10xCounts</a></span><span class="op">(</span><span class="st">"~/Share/scATAC/atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5"</span><span class="op">)</span></span>
<span><span class="va">scatac</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SingleCellExperiment 
dim: 89796 8728 
metadata(1): Samples
assays(1): counts
rownames(89796): chr1:565107-565550 chr1:569174-569639 ... chrY:59001782-59002175 chrY:59017143-59017246
rowData names(3): ID Symbol Type
colnames: NULL
colData names(2): Sample Barcode
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<p>This works because 10X Genomics make sure to distribute files in <code>.h5</code> format that are consistent accross single-cell sequencing methodologies.</p>
<p>However, the <code>SingleCellExperiment</code> obtained with this approach is not the most convenient, as it cannot natively leverage <code>fragments</code> file (see below).</p>
<p>Instead, we can create a <code>Signac</code> object, a flavour of <code>Seurat</code> objects.</p>
<div class="icon callout callout-style-default callout-question callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Import counts matrix and feature annotations using an import function provided by <code>Seurat</code>.</p>
<div class="icon callout callout-style-default callout-answer callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: SeuratObject</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: sp</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'SeuratObject'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    intersect, t</code></pre>
</div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac">Signac</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomicRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: stats4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: BiocGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: generics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'generics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    as.difftime, as.factor, as.ordered, intersect, is.element, setdiff, setequal, union</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind, colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep, grepl, is.unsorted, lapply, Map, mapply, match, mget, order, paste,
    pmax, pmax.int, pmin, pmin.int, Position, rank, rbind, Reduce, rownames, sapply, saveRDS, table, tapply, unique, unsplit, which.max, which.min</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: S4Vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'S4Vectors'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:utils':

    findMatches</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    expand.grid, I, unname</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: IRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'IRanges'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:sp':

    %over%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: Seqinfo</code></pre>
</div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org">stringr</a></span><span class="op">)</span></span>
<span><span class="va">cnts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X_h5.html">Read10X_h5</a></span><span class="op">(</span><span class="st">'~/Share/scATAC/atac_v1_pbmc_10k_filtered_peak_bc_matrix.h5'</span><span class="op">)</span></span>
<span><span class="va">features</span> <span class="op">&lt;-</span> <span class="fu">import</span><span class="op">(</span><span class="st">'~/Share/scATAC/atac_v1_pbmc_10k_peaks.bed'</span><span class="op">)</span></span>
<span><span class="va">features</span><span class="op">$</span><span class="va">peak</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">features</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html">str_replace</a></span><span class="op">(</span><span class="st">':'</span>, <span class="st">'-'</span><span class="op">)</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span></span>
<span>  file <span class="op">=</span> <span class="st">"~/Share/scATAC/atac_v1_pbmc_10k_singlecell.csv"</span>,</span>
<span>  header <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  row.names <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>How many accessible genomic segments were found in this dataset?</p>
<div class="icon callout callout-style-default callout-answer callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">features</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 89796 ranges and 1 metadata column:
          seqnames            ranges strand |                   peak
             &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |            &lt;character&gt;
      [1]     chr1     565108-565550      * |     chr1-565108-565550
      [2]     chr1     569175-569639      * |     chr1-569175-569639
      [3]     chr1     713461-714823      * |     chr1-713461-714823
      [4]     chr1     752423-753038      * |     chr1-752423-753038
      [5]     chr1     762107-763359      * |     chr1-762107-763359
      ...      ...               ...    ... .                    ...
  [89792]     chrY 58993393-58993760      * | chrY-58993393-58993760
  [89793]     chrY 58994572-58994823      * | chrY-58994572-58994823
  [89794]     chrY 58996353-58997331      * | chrY-58996353-58997331
  [89795]     chrY 59001783-59002175      * | chrY-59001783-59002175
  [89796]     chrY 59017144-59017246      * | chrY-59017144-59017246
  -------
  seqinfo: 24 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">features</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 89796</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="create-a-seurat-object" class="level3" data-number="13.1.2"><h3 data-number="13.1.2" class="anchored" data-anchor-id="create-a-seurat-object">
<span class="header-section-number">13.1.2</span> Create a <code>Seurat</code> object</h3>
<p>The next step is to aggregate <code>counts</code> and <code>features</code> into a <code>ChromatinAssay</code>, a scATAC-seq flavour of <code>Seurat</code> standard <code>Assays</code>. The documentation for <code><a href="https://stuartlab.org/signac/reference/CreateChromatinAssay.html">?CreateChromatinAssay</a></code> indicates that the user can provide:</p>
<ol type="1">
<li>A <code>fragments</code> file, corresponding to the full list of all unique fragments mapped across all single cells.</li>
<li>Genomic annotations to the <code>ChromatinAssay</code>, corresponding to gene annotations, promoter positions, etc. Such annotations can be generated from Ensembl.</li>
</ol>
<div class="icon callout callout-style-default callout-question callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Generate human annotations from Ensembl using a parsing function from <code>Seurat</code>.</p>
<div class="icon callout callout-style-default callout-answer callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-8-contents" aria-controls="callout-8" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-8" class="callout-8-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Get human gene annotations (hg19/GRCh37) to feed it into the future `ChromatinAssay`</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">'EnsDb.Hsapiens.v75'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>'getOption("repos")' replaces Bioconductor standard repositories, see 'help("repositories", package = "BiocManager")' for details.
Replacement repositories:
    CRAN: https://cloud.r-project.org</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Bioconductor version 3.22 (BiocManager 1.30.27), R 4.5.1 (2025-06-13)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package(s) not installed when version(s) same as or greater than current; use `force = TRUE` to re-install: 'EnsDb.Hsapiens.v75'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Installation paths not writeable, unable to update packages
  path: /opt/R/4.5.1/lib/R/library
  packages:
    boot, mgcv</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Old packages: 'affy', 'affyio', 'alphashape3d', 'AnnotationForge', 'AnnotationHubData', 'apeglm', 'aplot', 'arrow', 'aws.s3', 'AzureStor', 'basilisk.utils', 'BiocBook', 'BiocCheck', 'BiocPkgTools', 'biocthis', 'biocViews', 'biomformat',
  'broom.helpers', 'BrowserViz', 'bs4Dash', 'bumphunter', 'BumpyMatrix', 'cards', 'CGHbase', 'CGHcall', 'ChIPpeakAnno', 'ChIPseeker', 'clusterProfiler', 'collections', 'ComplexHeatmap', 'corrr', 'covr', 'crisprBase', 'crisprBowtie',
  'crisprScore', 'crisprScoreData', 'csaw', 'curatedMetagenomicData', 'curatedTCGAData', 'data.tree', 'dbscan', 'dearseq', 'DECIPHER', 'decontam', 'decoupleR', 'densvis', 'derfinder', 'derfinderHelper', 'DESeq2', 'DFplyr', 'dials',
  'diffHic', 'DNAcopy', 'DNAZooData', 'DOSE', 'DropletTestFiles', 'easylift', 'EBImage', 'ellmer', 'emmeans', 'EnrichedHeatmap', 'enrichplot', 'eulerr', 'excluderanges', 'ExperimentHubData', 'fgsea', 'fission', 'fourDNData', 'fresh',
  'genefilter', 'GenomeInfoDbData', 'GenomicFiles', 'GenomicInteractions', 'GEOquery', 'GGally', 'ggbio', 'ggExtra', 'ggimage', 'ggplotify', 'ggpubr', 'ggsci', 'ggstar', 'ggstats', 'ggtangle', 'ggtree', 'ggtreeExtra', 'gld', 'glmmTMB',
  'GO.db', 'GOSemSim', 'GOTHiC', 'gprofiler2', 'gsl', 'GSVA', 'gtrellis', 'GUniFrac', 'Gviz', 'hardhat', 'HCAData', 'HiCcompare', 'HiCDataLymphoblast', 'HiCDOC', 'HiCExperiment', 'HiContacts', 'HiContactsData', 'igvR', 'impute',
  'InteractionSet', 'interactiveDisplayBase', 'labelled', 'latticeExtra', 'lava', 'leaflet', 'lemon', 'LimROTS', 'litedown', 'lobstr', 'macrophage', 'magick', 'mariner', 'marray', 'MAST', 'Matrix', 'mclust', 'megadepth', 'memes',
  'MetaboCoreUtils', 'metagene2', 'mia', 'miaViz', 'microbiomeDataSets', 'MicrobiotaProcess', 'mixOmics', 'MOFA2', 'monaLisa', 'MotifDb', 'motifmatchr', 'msa', 'MsBackendMetaboLights', 'msdata', 'msmsEDA', 'msmsTests', 'MSnbase', 'MSnID',
  'MultiAssayExperiment', 'multiHiCcompare', 'muscat', 'mzID', 'mzR', 'NOISeq', 'nucleR', 'nullranges', 'oct4', 'ontoProc', 'ORFik', 'org.Hs.eg.db', 'org.Mm.eg.db', 'org.Sc.sgd.db', 'OrganismDbi', 'Orthology.eg.db', 'parsnip',
  'pcaMethods', 'pdftools', 'periodicDNA', 'philentropy', 'philr', 'phyloseq', 'PKI', 'plotgardener', 'plyinteractions', 'plyranges', 'pracma', 'preprocessCore', 'PSMatch', 'QDNAseq', 'QFeatures', 'quarto', 'qvalue', 'RaggedExperiment',
  'RBGL', 'Rbowtie', 'RcppParallel', 'rcrossref', 'REDseq', 'reformulas', 'regioneR', 'remaCor', 'Rfast', 'Rgraphviz', 'Rmpfr', 'ROC', 'rpx', 'rsconnect', 'rspm', 'rstatix', 'Rsubread', 'scatterpie', 'scmap', 'sechm', 'seriation',
  'shadowtext', 'shinydashboardPlus', 'ShortRead', 'singscore', 'snowflakeauth', 'SpatialExperiment', 'Spectra', 'spelling', 'SRAdb', 'styler', 'subSeq', 'survey', 'svglite', 'svUnit', 'TCC', 'terra', 'tidybulk', 'tidyCoverage',
  'tidySingleCellExperiment', 'tidySummarizedExperiment', 'timeDate', 'TMB', 'treeio', 'TreeSummarizedExperiment', 'tune', 'TxDb.Hsapiens.UCSC.hg19.knownGene', 'txdbmaker', 'tximeta', 'tximport', 'units', 'universalmotif',
  'variancePartition', 'vegan', 'VplotR', 'vsn', 'waiter', 'workflows', 'Wrench', 'yulab.utils', 'zinbwave'</code></pre>
</div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">annotations</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/GetGRangesFromEnsDb.html">GetGRangesFromEnsDb</a></span><span class="op">(</span>ensdb <span class="op">=</span> <span class="fu">EnsDb.Hsapiens.v75</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/EnsDb.Hsapiens.v75/man/package.html">EnsDb.Hsapiens.v75</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)
Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
  suppressWarnings() to suppress this warning.)</code></pre>
</div>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">GenomeInfoDb</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqlevelsStyle.html">seqlevelsStyle</a></span><span class="op">(</span><span class="va">annotations</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">'UCSC'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>Create a <code>ChromatinAssay</code> using <code>counts</code>, <code>features</code>, <code>fragments</code> and <code>annotations</code>.</p>
<div class="icon callout callout-style-default callout-answer callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Create Chromatin Assay</span></span>
<span><span class="va">assay</span> <span class="op">&lt;-</span> <span class="fu">Signac</span><span class="fu">::</span><span class="fu"><a href="https://stuartlab.org/signac/reference/CreateChromatinAssay.html">CreateChromatinAssay</a></span><span class="op">(</span></span>
<span>    counts <span class="op">=</span> <span class="va">cnts</span>,</span>
<span>    ranges <span class="op">=</span> <span class="va">features</span>,</span>
<span>    fragments <span class="op">=</span> <span class="st">"~/Share/scATAC/atac_v1_pbmc_10k_fragments.tsv.gz"</span>,</span>
<span>    annotation <span class="op">=</span> <span class="va">annotations</span>,</span>
<span>    genome <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>    min.cells <span class="op">=</span> <span class="fl">10</span>, </span>
<span>    min.features <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing hash</code></pre>
</div>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">assay</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ChromatinAssay data with 87561 features for 8728 cells
Variable features: 0 
Genome: hg19 
Annotation present: TRUE 
Motifs present: FALSE 
Fragment files: 1 </code></pre>
</div>
</div>
</div>
</div>
</div>
<p>What are the dimensions of this object? Are they comparable to the count matrix? Comment.</p>
<div class="icon callout callout-style-default callout-answer callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-10-contents" aria-controls="callout-10" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-10" class="callout-10-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 89796  8728</code></pre>
</div>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">assay</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 87561  8728</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<p>Itâ€™s finally time to wrap the <code>ChromatinAssay</code> into a <code>Seurat</code> standard object! This is done using the <code>CreateSeuratObject</code>, as already covered in <a href="../../content/day3/Lab6_batch_correction.html">Lab6</a></p>
<div class="icon callout callout-style-default callout-question callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Create a <code>Seurat</code> object from the <code>ChromatinAssay</code> object.</p>
<div class="icon callout callout-style-default callout-answer callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-12-contents" aria-controls="callout-12" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-12" class="callout-12-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Create Seurat object</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span></span>
<span>    counts <span class="op">=</span> <span class="va">assay</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">'ATAC'</span>, </span>
<span>    meta.data <span class="op">=</span> <span class="va">metadata</span><span class="op">[</span><span class="va">metadata</span><span class="op">$</span><span class="va">is__cell_barcode</span> <span class="op">==</span> <span class="fl">1</span>, <span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="va">PBMC</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
87561 features across 8728 samples within 1 assay 
Active assay: ATAC (87561 features, 0 variable features)
 2 layers present: counts, data</code></pre>
</div>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">PBMC</span><span class="op">[[</span><span class="st">'ATAC'</span><span class="op">]</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ChromatinAssay data with 87561 features for 8728 cells
Variable features: 0 
Genome: hg19 
Annotation present: TRUE 
Motifs present: FALSE 
Fragment files: 1 </code></pre>
</div>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html">granges</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 87561 ranges and 1 metadata column:
          seqnames            ranges strand |                   peak
             &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |            &lt;character&gt;
      [1]     chr1     565108-565550      * |     chr1-565108-565550
      [2]     chr1     569175-569639      * |     chr1-569175-569639
      [3]     chr1     713461-714823      * |     chr1-713461-714823
      [4]     chr1     752423-753038      * |     chr1-752423-753038
      [5]     chr1     762107-763359      * |     chr1-762107-763359
      ...      ...               ...    ... .                    ...
  [87557]     chrY 58993393-58993760      * | chrY-58993393-58993760
  [87558]     chrY 58994572-58994823      * | chrY-58994572-58994823
  [87559]     chrY 58996353-58997331      * | chrY-58996353-58997331
  [87560]     chrY 59001783-59002175      * | chrY-59001783-59002175
  [87561]     chrY 59017144-59017246      * | chrY-59017144-59017246
  -------
  seqinfo: 24 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://stuartlab.org/signac/reference/Annotation.html">Annotation</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3072120 ranges and 5 metadata columns:
                  seqnames        ranges strand |           tx_id   gene_name         gene_id   gene_biotype     type
                     &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt; &lt;character&gt;     &lt;character&gt;    &lt;character&gt; &lt;factor&gt;
  ENSE00001489430     chrX 192989-193061      + | ENST00000399012      PLCXD1 ENSG00000182378 protein_coding     exon
  ENSE00001536003     chrX 192991-193061      + | ENST00000484611      PLCXD1 ENSG00000182378 protein_coding     exon
  ENSE00002160563     chrX 193020-193061      + | ENST00000430923      PLCXD1 ENSG00000182378 protein_coding     exon
  ENSE00001750899     chrX 197722-197788      + | ENST00000445062      PLCXD1 ENSG00000182378 protein_coding     exon
  ENSE00001489388     chrX 197859-198351      + | ENST00000381657      PLCXD1 ENSG00000182378 protein_coding     exon
              ...      ...           ...    ... .             ...         ...             ...            ...      ...
  ENST00000361739    chrMT     7586-8269      + | ENST00000361739      MT-CO2 ENSG00000198712 protein_coding      cds
  ENST00000361789    chrMT   14747-15887      + | ENST00000361789      MT-CYB ENSG00000198727 protein_coding      cds
  ENST00000361851    chrMT     8366-8572      + | ENST00000361851     MT-ATP8 ENSG00000228253 protein_coding      cds
  ENST00000361899    chrMT     8527-9207      + | ENST00000361899     MT-ATP6 ENSG00000198899 protein_coding      cds
  ENST00000362079    chrMT     9207-9990      + | ENST00000362079      MT-CO3 ENSG00000198938 protein_coding      cds
  -------
  seqinfo: 25 sequences (1 circular) from hg19 genome</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="check-qcs" class="level3" data-number="13.1.3"><h3 data-number="13.1.3" class="anchored" data-anchor-id="check-qcs">
<span class="header-section-number">13.1.3</span> Check QCs</h3>
<section id="cell-based-qcs" class="level4" data-number="13.1.3.1"><h4 data-number="13.1.3.1" class="anchored" data-anchor-id="cell-based-qcs">
<span class="header-section-number">13.1.3.1</span> Cell-based QCs</h4>
<p>The fraction of reads in peaks (FRiP) is a good indicator of how well each cell was handled during scATACseq processing.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">PBMC</span><span class="op">$</span><span class="va">FRiP</span> <span class="op">&lt;-</span> <span class="va">PBMC</span><span class="op">$</span><span class="va">peak_region_fragments</span> <span class="op">/</span> <span class="va">PBMC</span><span class="op">$</span><span class="va">passed_filters</span></span>
<span><span class="va">PBMC</span><span class="op">$</span><span class="va">nCount_ATAC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/AssayData.html">GetAssayData</a></span><span class="op">(</span><span class="va">PBMC</span>, layer <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">PBMC</span><span class="op">$</span><span class="va">nFeature_ATAC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/AssayData.html">GetAssayData</a></span><span class="op">(</span><span class="va">PBMC</span>, layer <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">$</span><span class="va">FRiP</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>        0%        10%        20%        30%        40%        50%        60%        70%        80%        90%       100% 
0.02770289 0.55266699 0.60895480 0.64086129 0.66583010 0.68520756 0.70194678 0.71993709 0.73801044 0.76004660 0.90383142 </code></pre>
</div>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">$</span><span class="va">nCount_ATAC</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      0%      10%      20%      30%      40%      50%      60%      70%      80%      90%     100% 
  1710.0   5079.1   7619.8   9522.5  11300.8  13367.5  15754.6  19005.5  23502.6  31316.2 249799.0 </code></pre>
</div>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">$</span><span class="va">nFeature_ATAC</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     0%     10%     20%     30%     40%     50%     60%     70%     80%     90%    100% 
  654.0  2308.4  3349.4  4108.1  4765.0  5447.0  6234.2  7100.9  8241.6  9983.4 30313.0 </code></pre>
</div>
</div>
</section><section id="peaks-based-qcs" class="level4" data-number="13.1.3.2"><h4 data-number="13.1.3.2" class="anchored" data-anchor-id="peaks-based-qcs">
<span class="header-section-number">13.1.3.2</span> Peaks-based QCs</h4>
<div class="icon callout callout-style-default callout-question callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Which analysis are the <code>fragments</code> required for, exactly?</p>
<p>Could we still perform normalization/clustering/annotation without them? And motif enrichment analysis?</p>
<p>Since we do have the <code>fragments</code> file at hand, most of the QC steps are available (e.g.&nbsp;<code>TSSEnrichment</code>, <code>NucleosomeSignal</code> or fragment size distribution). Letâ€™s go through them one by one.</p>
<div class="icon callout callout-style-default callout-answer callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-14-contents" aria-controls="callout-14" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-14" class="callout-14-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># compute nucleosome signal score per cell</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/NucleosomeSignal.html">NucleosomeSignal</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">PBMC</span><span class="op">)</span></span>
<span><span class="va">PBMC</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
87561 features across 8728 samples within 1 assay 
Active assay: ATAC (87561 features, 0 variable features)
 2 layers present: counts, data</code></pre>
</div>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># compute TSS enrichment score per cell</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu">Signac</span><span class="fu">::</span><span class="fu"><a href="https://stuartlab.org/signac/reference/TSSEnrichment.html">TSSEnrichment</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">PBMC</span>, fast <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Extracting TSS positions</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding + strand cut sites</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding - strand cut sites</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing mean insertion frequency in flanking regions</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing TSS score</code></pre>
</div>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">PBMC</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
87561 features across 8728 samples within 1 assay 
Active assay: ATAC (87561 features, 0 variable features)
 2 layers present: counts, data</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
<p>The <code>TSSPlot</code> function from <code>Signac</code> can be used to plot the fragment count per peak ~ TSS enrichment.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">PBMC</span><span class="op">$</span><span class="va">high.tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">$</span><span class="va">TSS.enrichment</span> <span class="op">&gt;</span> <span class="fl">3.5</span>, <span class="st">'High'</span>, <span class="st">'Low'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://stuartlab.org/signac/reference/TSSPlot.html">TSSPlot</a></span><span class="op">(</span><span class="va">PBMC</span>, group.by <span class="op">=</span> <span class="st">'high.tss'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
â„¹ Please use `linewidth` instead.
â„¹ The deprecated feature was likely used in the Signac package.
  Please report the issue at &lt;https://github.com/stuart-lab/signac/issues&gt;.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="Lab7_atac-seq_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">PBMC</span><span class="op">$</span><span class="va">high.tss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">$</span><span class="va">TSS.enrichment</span> <span class="op">&gt;</span> <span class="fl">2.5</span>, <span class="st">'High'</span>, <span class="st">'Low'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://stuartlab.org/signac/reference/TSSPlot.html">TSSPlot</a></span><span class="op">(</span><span class="va">PBMC</span>, group.by <span class="op">=</span> <span class="st">'high.tss'</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="Lab7_atac-seq_files/figure-html/unnamed-chunk-11-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The <code>FragmentHistogram</code> function from <code>Signac</code> can be used to plot the fragment size distribution in peaks with different nucleosome signals.</p>
<div class="icon callout callout-style-default callout-question callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Group genomic loci by their nucleosome signal (&gt; 4 or &lt; 4) and plot the fragment size distribution.</p>
<div class="icon callout callout-style-default callout-answer callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-16-contents" aria-controls="callout-16" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-16" class="callout-16-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">PBMC</span><span class="op">$</span><span class="va">nucleosome_group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">$</span><span class="va">nucleosome_signal</span> <span class="op">&gt;</span> <span class="fl">4</span>, <span class="st">'NS &gt; 4'</span>, <span class="st">'NS &lt; 4'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://stuartlab.org/signac/reference/FragmentHistogram.html">FragmentHistogram</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">PBMC</span>, group.by <span class="op">=</span> <span class="st">'nucleosome_group'</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 92 rows containing non-finite outside the scale range (`stat_bin()`).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 4 rows containing missing values or values outside the scale range (`geom_bar()`).</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="Lab7_atac-seq_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section></section><section id="filter-cells-and-features" class="level3" data-number="13.1.4"><h3 data-number="13.1.4" class="anchored" data-anchor-id="filter-cells-and-features">
<span class="header-section-number">13.1.4</span> Filter cells and features</h3>
<p>Just like standard scRNAseq data, scATACseq data needs to be filtered to remove low-quality cells and peaks.</p>
<div class="icon callout callout-style-default callout-question callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Filter the <code>Seurat</code> object (cells and features) as deemed appropriate.</p>
<div class="icon callout callout-style-default callout-answer callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-18-contents" aria-controls="callout-18" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-18" class="callout-18-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Filter data </span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">PBMC</span>, subset <span class="op">=</span> <span class="va">nCount_ATAC</span> <span class="op">&gt;</span> <span class="fl">3000</span> <span class="op">&amp;</span> <span class="va">nCount_ATAC</span> <span class="op">&lt;</span> <span class="fl">100000</span><span class="op">)</span> </span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">PBMC</span>, subset <span class="op">=</span> <span class="va">nFeature_ATAC</span> <span class="op">&gt;</span> <span class="fl">1000</span> <span class="op">&amp;</span> <span class="va">nFeature_ATAC</span> <span class="op">&lt;</span> <span class="fl">20000</span><span class="op">)</span> </span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">PBMC</span>, subset <span class="op">=</span> <span class="va">FRiP</span> <span class="op">&gt;</span> <span class="fl">0.30</span><span class="op">)</span> </span>
<span><span class="va">PBMC</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
87561 features across 8231 samples within 1 assay 
Active assay: ATAC (87561 features, 0 variable features)
 2 layers present: counts, data</code></pre>
</div>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Remove peaks with low coverage</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="va">PBMC</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/AssayData.html">GetAssayData</a></span><span class="op">(</span><span class="va">PBMC</span>, layer <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">10</span>, <span class="op">]</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="va">PBMC</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/AssayData.html">GetAssayData</a></span><span class="op">(</span><span class="va">PBMC</span>, layer <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">10</span>, <span class="op">]</span></span>
<span><span class="va">PBMC</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
87212 features across 8231 samples within 1 assay 
Active assay: ATAC (87212 features, 0 variable features)
 2 layers present: counts, data</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="dimensionality-reduction-and-clustering" class="level3" data-number="13.1.5"><h3 data-number="13.1.5" class="anchored" data-anchor-id="dimensionality-reduction-and-clustering">
<span class="header-section-number">13.1.5</span> Dimensionality reduction and clustering</h3>
<p>Now that the dataset is filtered, we can proceed to data normalization, dimensionality reduction and clustering. Compared to scRNAseq, scATACseq data is much sparser and has a higher dimensionality. This has important implications for the choice of normalization and dimensionality reduction methods. scATACseq data are generally normalized using a <code>TF-IDF</code> normalization, which is implemented in the <code>RunTFIDF</code> function in <code>Signac</code>, and dimensionality reduction is performed using <code>LSI</code> (Latent Semantic Indexing) implemented in the <code>RunSVD</code> function.</p>
<div class="icon callout callout-style-default callout-question callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Normalize then further reduce the dimensionality for visualization purposes.</p>
<div class="icon callout callout-style-default callout-answer callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-20-contents" aria-controls="callout-20" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-20" class="callout-20-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Normalize data </span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/RunTFIDF.html">RunTFIDF</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Performing TF-IDF normalization</code></pre>
</div>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Reduce dimensionality</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/FindTopFeatures.html">FindTopFeatures</a></span><span class="op">(</span><span class="va">PBMC</span>, min.cutoff <span class="op">=</span> <span class="st">'q50'</span><span class="op">)</span> </span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/RunSVD.html">RunSVD</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Running SVD</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Scaling cell embeddings</code></pre>
</div>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Embed in 2D UMAP</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">PBMC</span>, reduction <span class="op">=</span> <span class="st">'lsi'</span>, dims <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
This message will be shown once per session</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>20:24:04 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>20:24:04 Read 8231 rows and found 29 numeric columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>20:24:04 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>20:24:04 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>[----|----|----|----|----|----|----|----|----|----|</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**************************************************|
20:24:05 Writing NN index file to temp file /data/.cache/R//RtmpaFe87f/file16a08c51315069
20:24:05 Searching Annoy index using 1 thread, search_k = 3000
20:24:07 Annoy recall = 100%
20:24:08 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30
20:24:09 Initializing from normalized Laplacian + noise (using RSpectra)
20:24:09 Commencing optimization for 500 epochs, with 316686 positive edges
20:24:09 Using rng type: pcg
20:24:16 Optimization finished</code></pre>
</div>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">## - Cluster cells</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">PBMC</span>, reduction <span class="op">=</span> <span class="st">'lsi'</span>, dims <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing nearest neighbor graph
Computing SNN</code></pre>
</div>
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">PBMC</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>, algorithm <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## - Visualize data </span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">PBMC</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="Lab7_atac-seq_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<p>What can you observe in the <code>UMAP</code> projection of the dataset? Comment on the separation of some cell types into different spatially-resolved clusters.</p>
</div>
</div>
</section></section><section id="compute-gene-activity-scores" class="level2" data-number="13.2"><h2 data-number="13.2" class="anchored" data-anchor-id="compute-gene-activity-scores">
<span class="header-section-number">13.2</span> Compute gene activity scores</h2>
<p>Signacâ€™s <code><a href="https://stuartlab.org/signac/reference/GeneActivity.html">GeneActivity()</a></code> function require scATACseq fragment information. Since we have them, we can estimate a gene activity score for each gene in the <code>annotations</code>.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">gene.activities</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/GeneActivity.html">GeneActivity</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Extracting gene coordinates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in SingleFeatureMatrix(fragment = fragments[[x]], features = features, : 13 features are on seqnames not present in the fragment file. These will be removed.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Extracting reads overlapping genomic regions</code></pre>
</div>
</div>
<p>We can now save this new object as an <code>Assay</code> in the <code>PBMC</code> object and normalize it.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">PBMC</span><span class="op">[[</span><span class="st">'RNA'</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/CreateAssayObject.html">CreateAssayObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">gene.activities</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Non-unique features (rownames) present in the input matrix, making unique</code></pre>
</div>
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># - Normalize the new RNA assay.</span></span>
<span><span class="va">PBMC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">PBMC</span>, assay <span class="op">=</span> <span class="st">'RNA'</span>, normalization.method <span class="op">=</span> <span class="st">'LogNormalize'</span>, scale.factor <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">$</span><span class="va">nCount_RNA</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">PBMC</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
107222 features across 8231 samples within 2 assays 
Active assay: ATAC (87212 features, 43610 variable features)
 2 layers present: counts, data
 1 other assay present: RNA
 2 dimensional reductions calculated: lsi, umap</code></pre>
</div>
</div>
<p>One can now visualize expression of individual markers across clusters to infer cluster identity.</p>
<div class="icon callout callout-style-default callout-question callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Question
</div>
</div>
<div class="callout-body-container callout-body">
<p>Assess the expression of known markers across clusters, projecting the cells in a 2D UMAP.</p>
<div class="icon callout callout-style-default callout-answer callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-22-contents" aria-controls="callout-22" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-22" class="callout-22-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">PBMC</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span></span>
<span>    <span class="st">'MS4A1'</span>, <span class="co"># B-cell</span></span>
<span>    <span class="st">'CD3D'</span>, <span class="co"># T-cell</span></span>
<span>    <span class="st">'NKG7'</span>, <span class="co"># NK cell</span></span>
<span>    <span class="st">'TREM1'</span> <span class="co"># Monocytes</span></span>
<span>  <span class="op">)</span>, </span>
<span>  assay <span class="op">=</span> <span class="st">"RNA"</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="Lab7_atac-seq_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">PBMC</span><span class="op">$</span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html">case_when</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5</span>, <span class="fl">9</span>, <span class="fl">12</span><span class="op">)</span> <span class="op">~</span> <span class="st">"Monocytes"</span>, </span>
<span>  <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">7</span>, <span class="fl">10</span><span class="op">)</span> <span class="op">~</span> <span class="st">"T-cells"</span>, </span>
<span>  <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">6</span><span class="op">)</span> <span class="op">~</span> <span class="st">"NK-cells"</span>, </span>
<span>  <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">8</span>, <span class="fl">11</span><span class="op">)</span> <span class="op">~</span> <span class="st">"B-cells"</span>, </span>
<span>  <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">13</span>, <span class="fl">14</span><span class="op">)</span> <span class="op">~</span> <span class="st">"DC"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">PBMC</span>, group.by <span class="op">=</span> <span class="st">'annotation'</span>, label <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="Lab7_atac-seq_files/figure-html/unnamed-chunk-17-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="find-peaks-differentially-accessible-across-clusters" class="level2" data-number="13.3"><h2 data-number="13.3" class="anchored" data-anchor-id="find-peaks-differentially-accessible-across-clusters">
<span class="header-section-number">13.3</span> Find peaks differentially accessible across clusters</h2>
<p>Finally, we can perform differential peak accessibility analysis to identify peaks that are differentially accessible between clusters. This is done using the <code>FindMarkers</code> function in <code>Signac</code>, which is similar to <code>FindAllMarkers</code> in <code>Seurat</code>.</p>
<div class="cell">
<div class="code-with-filename">
<div class="code-with-filename-file">
<pre><strong>R</strong></pre>
</div>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">PBMC</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">PBMC</span><span class="op">$</span><span class="va">annotation</span></span>
<span><span class="va">da_peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">PBMC</span>,</span>
<span>  ident.1 <span class="op">=</span> <span class="st">"Monocytes"</span>,</span>
<span>  ident.2 <span class="op">=</span> <span class="st">"B-cells"</span>,</span>
<span>  test.use <span class="op">=</span> <span class="st">'LR'</span>,</span>
<span>  latent.vars <span class="op">=</span> <span class="st">'nCount_ATAC'</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span></span>
<span>  object <span class="op">=</span> <span class="va">PBMC</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">da_peaks</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>  pt.size <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  idents <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Monocytes"</span>,<span class="st">"B-cells"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section><section id="session-info" class="level2" data-number="13.4"><h2 data-number="13.4" class="anchored" data-anchor-id="session-info">
<span class="header-section-number">13.4</span> Session info</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>â”€ Session info â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
 setting  value
 version  R version 4.5.1 (2025-06-13)
 os       Ubuntu 24.04.2 LTS
 system   x86_64, linux-gnu
 ui       X11
 language (EN)
 collate  en_IN.utf8
 ctype    en_IN.utf8
 tz       Europe/Paris
 date     2025-11-14
 pandoc   3.1.3 @ /usr/bin/ (via rmarkdown)
 quarto   1.5.55 @ /opt/quarto/bin/quarto

â”€ Packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€
 package              * version   date (UTC) lib source
 abind                  1.4-8     2024-09-12 [1] CRAN (R 4.5.1)
 AnnotationDbi          1.72.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 AnnotationFilter       1.34.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 backports              1.5.0     2024-05-23 [1] CRAN (R 4.5.1)
 base64enc              0.1-3     2015-07-28 [1] CRAN (R 4.5.1)
 beachmat               2.26.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 beeswarm               0.4.0     2021-06-01 [1] CRAN (R 4.5.1)
 Biobase                2.70.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 BiocGenerics         * 0.56.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 BiocIO                 1.20.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 BiocManager            1.30.27   2025-11-14 [1] CRAN (R 4.5.1)
 BiocParallel           1.44.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 Biostrings             2.78.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 biovizBase             1.58.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 bit                    4.6.0     2025-03-06 [1] CRAN (R 4.5.0)
 bit64                  4.6.0-1   2025-01-16 [1] CRAN (R 4.5.0)
 bitops                 1.0-9     2024-10-03 [1] CRAN (R 4.5.1)
 blob                   1.2.4     2023-03-17 [1] CRAN (R 4.5.1)
 BSgenome               1.78.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 cachem                 1.1.0     2024-05-16 [1] CRAN (R 4.5.1)
 checkmate              2.3.3     2025-08-18 [1] CRAN (R 4.5.1)
 cigarillo              1.0.0     2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 cli                    3.6.5     2025-04-23 [1] CRAN (R 4.5.1)
 cluster                2.1.8.1   2025-03-12 [2] CRAN (R 4.5.1)
 codetools              0.2-20    2024-03-31 [2] CRAN (R 4.5.1)
 colorspace             2.1-2     2025-09-22 [1] CRAN (R 4.5.1)
 cowplot                1.2.0     2025-07-07 [1] CRAN (R 4.5.1)
 crayon                 1.5.3     2024-06-20 [1] CRAN (R 4.5.1)
 curl                   7.0.0     2025-08-19 [1] CRAN (R 4.5.1)
 data.table             1.17.8    2025-07-10 [1] CRAN (R 4.5.1)
 DBI                    1.2.3     2024-06-02 [1] CRAN (R 4.5.1)
 DelayedArray           0.36.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 DelayedMatrixStats     1.32.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 deldir                 2.0-4     2024-02-28 [1] CRAN (R 4.5.1)
 devtools               2.4.6     2025-10-03 [1] CRAN (R 4.5.1)
 dichromat              2.0-0.1   2022-05-02 [1] CRAN (R 4.5.1)
 digest                 0.6.38    2025-11-12 [1] CRAN (R 4.5.1)
 dotCall64              1.2       2024-10-04 [1] CRAN (R 4.5.1)
 dplyr                  1.1.4     2023-11-17 [1] CRAN (R 4.5.1)
 dqrng                  0.4.1     2024-05-28 [1] CRAN (R 4.5.1)
 DropletUtils           1.30.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 edgeR                  4.8.0     2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 ellipsis               0.3.2     2021-04-29 [1] CRAN (R 4.5.1)
 EnsDb.Hsapiens.v75     2.99.0    2025-07-16 [1] Bioconductor
 ensembldb              2.34.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 evaluate               1.0.5     2025-08-27 [1] CRAN (R 4.5.1)
 farver                 2.1.2     2024-05-13 [1] CRAN (R 4.5.1)
 fastDummies            1.7.5     2025-01-20 [1] CRAN (R 4.5.0)
 fastmap                1.2.0     2024-05-15 [1] CRAN (R 4.5.1)
 fastmatch              1.1-6     2024-12-23 [1] CRAN (R 4.5.0)
 fitdistrplus           1.2-4     2025-07-03 [1] CRAN (R 4.5.1)
 foreign                0.8-90    2025-03-31 [2] CRAN (R 4.5.1)
 Formula                1.2-5     2023-02-24 [1] CRAN (R 4.5.1)
 fs                     1.6.6     2025-04-12 [1] CRAN (R 4.5.1)
 future               * 1.67.0    2025-07-29 [1] CRAN (R 4.5.1)
 future.apply           1.20.0    2025-06-06 [1] CRAN (R 4.5.1)
 generics             * 0.1.4     2025-05-09 [1] CRAN (R 4.5.1)
 GenomeInfoDb           1.46.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 GenomicAlignments      1.46.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 GenomicFeatures        1.62.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 GenomicRanges        * 1.62.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 ggbeeswarm             0.7.2     2023-04-29 [1] CRAN (R 4.5.1)
 ggplot2                4.0.1     2025-11-14 [1] CRAN (R 4.5.1)
 ggrastr                1.0.2     2023-06-01 [1] CRAN (R 4.5.1)
 ggrepel                0.9.6     2024-09-07 [1] CRAN (R 4.5.1)
 ggridges               0.5.7     2025-08-27 [1] CRAN (R 4.5.1)
 globals                0.18.0    2025-05-08 [1] CRAN (R 4.5.1)
 glue                   1.8.0     2024-09-30 [1] CRAN (R 4.5.1)
 goftest                1.2-3     2021-10-07 [1] CRAN (R 4.5.1)
 gridExtra              2.3       2017-09-09 [1] CRAN (R 4.5.1)
 gtable                 0.3.6     2024-10-25 [1] CRAN (R 4.5.1)
 h5mread                1.2.0     2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 HDF5Array              1.38.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 hdf5r                  1.3.12    2025-01-20 [1] CRAN (R 4.5.1)
 Hmisc                  5.2-4     2025-10-05 [1] CRAN (R 4.5.1)
 htmlTable              2.4.3     2024-07-21 [1] CRAN (R 4.5.1)
 htmltools              0.5.8.1   2024-04-04 [1] CRAN (R 4.5.1)
 htmlwidgets            1.6.4     2023-12-06 [1] CRAN (R 4.5.1)
 httpuv                 1.6.16    2025-04-16 [1] CRAN (R 4.5.1)
 httr                   1.4.7     2023-08-15 [1] CRAN (R 4.5.1)
 ica                    1.0-3     2022-07-08 [1] CRAN (R 4.5.1)
 igraph                 2.2.1     2025-10-27 [1] CRAN (R 4.5.1)
 IRanges              * 2.44.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 irlba                  2.3.5.1   2022-10-03 [1] CRAN (R 4.5.1)
 jsonlite               2.0.0     2025-03-27 [1] CRAN (R 4.5.1)
 KEGGREST               1.50.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 KernSmooth             2.23-26   2025-01-01 [2] CRAN (R 4.5.1)
 knitr                  1.50      2025-03-16 [1] CRAN (R 4.5.1)
 labeling               0.4.3     2023-08-29 [1] CRAN (R 4.5.1)
 later                  1.4.4     2025-08-27 [1] CRAN (R 4.5.1)
 lattice                0.22-7    2025-04-02 [2] CRAN (R 4.5.1)
 lazyeval               0.2.2     2019-03-15 [1] CRAN (R 4.5.1)
 lifecycle              1.0.4     2023-11-07 [1] CRAN (R 4.5.1)
 limma                  3.66.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 listenv                0.10.0    2025-11-02 [1] CRAN (R 4.5.1)
 lmtest                 0.9-40    2022-03-21 [1] CRAN (R 4.5.1)
 locfit                 1.5-9.12  2025-03-05 [1] CRAN (R 4.5.0)
 magrittr               2.0.4     2025-09-12 [1] CRAN (R 4.5.1)
 MASS                   7.3-65    2025-02-28 [2] CRAN (R 4.5.1)
 Matrix                 1.7-3     2025-03-11 [1] CRAN (R 4.5.1)
 MatrixGenerics         1.22.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 matrixStats            1.5.0     2025-01-07 [1] CRAN (R 4.5.0)
 memoise                2.0.1     2021-11-26 [1] CRAN (R 4.5.1)
 mime                   0.13      2025-03-17 [1] CRAN (R 4.5.1)
 miniUI                 0.1.2     2025-04-17 [1] CRAN (R 4.5.1)
 nlme                   3.1-168   2025-03-31 [2] CRAN (R 4.5.1)
 nnet                   7.3-20    2025-01-01 [2] CRAN (R 4.5.1)
 otel                   0.2.0     2025-08-29 [1] CRAN (R 4.5.1)
 parallelly             1.45.1    2025-07-24 [1] CRAN (R 4.5.1)
 patchwork              1.3.2     2025-08-25 [1] CRAN (R 4.5.1)
 pbapply                1.7-4     2025-07-20 [1] CRAN (R 4.5.1)
 pillar                 1.11.1    2025-09-17 [1] CRAN (R 4.5.1)
 pkgbuild               1.4.8     2025-05-26 [1] CRAN (R 4.5.1)
 pkgconfig              2.0.3     2019-09-22 [1] CRAN (R 4.5.1)
 pkgload                1.4.1     2025-09-23 [1] CRAN (R 4.5.1)
 plotly                 4.11.0    2025-06-19 [1] CRAN (R 4.5.1)
 plyr                   1.8.9     2023-10-02 [1] CRAN (R 4.5.1)
 png                    0.1-8     2022-11-29 [1] CRAN (R 4.5.1)
 polyclip               1.10-7    2024-07-23 [1] CRAN (R 4.5.1)
 progressr              0.18.0    2025-11-06 [1] CRAN (R 4.5.1)
 promises               1.5.0     2025-11-01 [1] CRAN (R 4.5.1)
 ProtGenerics           1.42.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 purrr                  1.2.0     2025-11-04 [1] CRAN (R 4.5.1)
 R.methodsS3            1.8.2     2022-06-13 [1] CRAN (R 4.5.1)
 R.oo                   1.27.1    2025-05-02 [1] CRAN (R 4.5.1)
 R.utils                2.13.0    2025-02-24 [1] CRAN (R 4.5.0)
 R6                     2.6.1     2025-02-15 [1] CRAN (R 4.5.0)
 RANN                   2.6.2     2024-08-25 [1] CRAN (R 4.5.1)
 RColorBrewer           1.1-3     2022-04-03 [1] CRAN (R 4.5.1)
 Rcpp                   1.1.0     2025-07-02 [1] CRAN (R 4.5.1)
 RcppAnnoy              0.0.22    2024-01-23 [1] CRAN (R 4.5.1)
 RcppHNSW               0.6.0     2024-02-04 [1] CRAN (R 4.5.1)
 RcppRoll               0.3.1     2024-07-07 [1] CRAN (R 4.5.1)
 RCurl                  1.98-1.17 2025-03-22 [1] CRAN (R 4.5.1)
 remotes                2.5.0     2024-03-17 [1] CRAN (R 4.5.1)
 reshape2               1.4.5     2025-11-12 [1] CRAN (R 4.5.1)
 restfulr               0.0.16    2025-06-27 [1] CRAN (R 4.5.1)
 reticulate             1.44.0    2025-10-25 [1] CRAN (R 4.5.1)
 rhdf5                  2.54.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 rhdf5filters           1.22.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 Rhdf5lib               1.32.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 rjson                  0.2.23    2024-09-16 [1] CRAN (R 4.5.1)
 rlang                  1.1.6     2025-04-11 [1] CRAN (R 4.5.1)
 rmarkdown              2.30      2025-09-28 [1] CRAN (R 4.5.1)
 ROCR                   1.0-11    2020-05-02 [1] CRAN (R 4.5.1)
 rpart                  4.1.24    2025-01-07 [2] CRAN (R 4.5.1)
 Rsamtools              2.26.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 RSpectra               0.16-2    2024-07-18 [1] CRAN (R 4.5.1)
 RSQLite                2.4.4     2025-11-10 [1] CRAN (R 4.5.1)
 rstudioapi             0.17.1    2024-10-22 [1] CRAN (R 4.5.1)
 rtracklayer          * 1.70.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 Rtsne                  0.17      2023-12-07 [1] CRAN (R 4.5.1)
 S4Arrays               1.10.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 S4Vectors            * 0.48.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 S7                     0.2.0     2024-11-07 [1] CRAN (R 4.5.1)
 scales                 1.4.0     2025-04-24 [1] CRAN (R 4.5.1)
 scattermore            1.2       2023-06-12 [1] CRAN (R 4.5.1)
 sctransform            0.4.2     2025-04-30 [1] CRAN (R 4.5.1)
 scuttle                1.20.0    2025-10-30 [1] Bioconductor 3.22 (R 4.5.1)
 Seqinfo              * 1.0.0     2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 sessioninfo            1.2.3     2025-02-05 [1] CRAN (R 4.5.0)
 Seurat               * 5.3.1     2025-10-29 [1] CRAN (R 4.5.1)
 SeuratObject         * 5.2.0     2025-08-27 [1] CRAN (R 4.5.1)
 shiny                  1.11.1    2025-07-03 [1] CRAN (R 4.5.1)
 Signac               * 1.16.0    2025-10-10 [1] CRAN (R 4.5.1)
 SingleCellExperiment   1.32.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 sp                   * 2.2-0     2025-02-01 [1] CRAN (R 4.5.0)
 spam                   2.11-1    2025-01-20 [1] CRAN (R 4.5.0)
 SparseArray            1.10.1    2025-10-31 [1] Bioconductor 3.22 (R 4.5.1)
 sparseMatrixStats      1.22.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 spatstat.data          3.1-9     2025-10-18 [1] CRAN (R 4.5.1)
 spatstat.explore       3.5-3     2025-09-22 [1] CRAN (R 4.5.1)
 spatstat.geom          3.6-0     2025-09-20 [1] CRAN (R 4.5.1)
 spatstat.random        3.4-2     2025-09-21 [1] CRAN (R 4.5.1)
 spatstat.sparse        3.1-0     2024-06-21 [1] CRAN (R 4.5.1)
 spatstat.univar        3.1-4     2025-07-13 [1] CRAN (R 4.5.1)
 spatstat.utils         3.2-0     2025-09-20 [1] CRAN (R 4.5.1)
 statmod                1.5.1     2025-10-09 [1] CRAN (R 4.5.1)
 stringi                1.8.7     2025-03-27 [1] CRAN (R 4.5.1)
 stringr              * 1.6.0     2025-11-04 [1] CRAN (R 4.5.1)
 SummarizedExperiment   1.40.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 survival               3.8-3     2024-12-17 [2] CRAN (R 4.5.1)
 tensor                 1.5.1     2025-06-17 [1] CRAN (R 4.5.1)
 tibble                 3.3.0     2025-06-08 [1] CRAN (R 4.5.1)
 tidyr                  1.3.1     2024-01-24 [1] CRAN (R 4.5.1)
 tidyselect             1.2.1     2024-03-11 [1] CRAN (R 4.5.1)
 UCSC.utils             1.6.0     2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 usethis                3.2.1     2025-09-06 [1] CRAN (R 4.5.1)
 uwot                   0.2.4     2025-11-10 [1] CRAN (R 4.5.1)
 VariantAnnotation      1.56.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 vctrs                  0.6.5     2023-12-01 [1] CRAN (R 4.5.1)
 vipor                  0.4.7     2023-12-18 [1] CRAN (R 4.5.1)
 viridisLite            0.4.2     2023-05-02 [1] CRAN (R 4.5.1)
 withr                  3.0.2     2024-10-28 [1] CRAN (R 4.5.1)
 xfun                   0.54      2025-10-30 [1] CRAN (R 4.5.1)
 XML                    3.99-0.20 2025-11-08 [1] CRAN (R 4.5.1)
 xtable                 1.8-4     2019-04-21 [1] CRAN (R 4.5.1)
 XVector                0.50.0    2025-10-29 [1] Bioconductor 3.22 (R 4.5.1)
 yaml                   2.3.10    2024-07-26 [1] CRAN (R 4.5.1)
 zoo                    1.8-14    2025-04-10 [1] CRAN (R 4.5.1)

 [1] /home/rsg/R/x86_64-pc-linux-gnu-library/4.5
 [2] /opt/R/4.5.1/lib/R/library
 * â”€â”€ Packages attached to the search path.

â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</code></pre>
</div>
</div>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/js2264\.github\.io\/scRNAseq_Physalia_2025\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="../../content/day4/Lecture6_ATAC.html" class="pagination-link" aria-label="Lecture 6 - Advances in single-cell genomics: the epigenome">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Lecture 6 - Advances in single-cell genomics: the epigenome</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../content/day4/Lecture7_pseudotime.html" class="pagination-link" aria-label="Lecture 7 - Trajectories and pseudotimes">
        <span class="nav-page-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Lecture 7 - Trajectories and pseudotimes</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p>Single-cell RNAseq analysis with R/Bioconductor |<br>
J. Serizay, O. Ashenberg, F. Almeida-Silva</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    <div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/js2264/scRNAseq_Physalia_2025/edit/main/content/day4/Lab7_atac-seq.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/js2264/scRNAseq_Physalia_2025/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
<p>This book was built with <a href="https://quarto.org/">Quarto</a>.</p>
</div>
  </div>
</footer>


</body></html>